// @allianz/ngx-ndbx/esm2015/progress-stepper/buttons.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CdkStepper, CdkStepperNext, CdkStepperPrevious } from '@angular/cdk/stepper';
import { ChangeDetectorRef, Directive } from '@angular/core';
import { NxProgressStepperDirective } from './progress-stepper.component';
// tslint:disable:use-input-property-decorator
export class NxStepperNextDirective extends CdkStepperNext {
    /**
     * @param {?} _stepper
     * @param {?} changeDetectorRef
     */
    constructor(_stepper, changeDetectorRef) {
        super(_stepper);
        this.changeDetectorRef = changeDetectorRef;
        /**
         * \@docs-private
         */
        this.disabled = false;
    }
    /**
     * Implemented to prevent changed after checked error after stepper init.
     * When the stepper initializes the directive is checked first before
     * the stepper can check its content thus it doesn't know about it's children yet
     * so _stepper.hasNext returns false and disables the buttons. In the same CD cycle
     * after the stepper has checked its contents the button needs to be enabled resulting
     * in the error.
     * @return {?}
     */
    ngDoCheck() {
        /** @type {?} */
        const stepper = (/** @type {?} */ (this._stepper));
        if (this.disabled !== !stepper.hasNext) {
            this.disabled = !stepper.hasNext;
            this.changeDetectorRef.markForCheck();
        }
    }
}
NxStepperNextDirective.decorators = [
    { type: Directive, args: [{
                selector: 'button[nxStepperNext]',
                host: {
                    '(click)': '_stepper.next()',
                    '[type]': 'type',
                    '[disabled]': 'disabled'
                },
                inputs: ['type'],
                providers: [{ provide: CdkStepper, useExisting: NxProgressStepperDirective }]
            },] }
];
/** @nocollapse */
NxStepperNextDirective.ctorParameters = () => [
    { type: CdkStepper },
    { type: ChangeDetectorRef }
];
if (false) {
    /**
     * \@docs-private
     * @type {?}
     */
    NxStepperNextDirective.prototype.disabled;
    /**
     * @type {?}
     * @private
     */
    NxStepperNextDirective.prototype.changeDetectorRef;
}
export class NxStepperPreviousDirective extends CdkStepperPrevious {
    /**
     * @param {?} _stepper
     * @param {?} changeDetectorRef
     */
    constructor(_stepper, changeDetectorRef) {
        super(_stepper);
        this.changeDetectorRef = changeDetectorRef;
    }
}
NxStepperPreviousDirective.decorators = [
    { type: Directive, args: [{
                selector: 'button[nxStepperPrevious]',
                host: {
                    '(click)': '_stepper.previous()',
                    '[type]': 'type',
                    '[disabled]': '!_stepper.hasPrevious'
                },
                inputs: ['type'],
                providers: [{ provide: CdkStepper, useExisting: NxProgressStepperDirective }]
            },] }
];
/** @nocollapse */
NxStepperPreviousDirective.ctorParameters = () => [
    { type: CdkStepper },
    { type: ChangeDetectorRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxStepperPreviousDirective.prototype.changeDetectorRef;
}
