// @allianz/ngx-ndbx/esm2015/page-search/page-search.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, Output } from '@angular/core';
export class NxPageSearchComponent {
    /**
     * @param {?} _changeDetectorRef
     */
    constructor(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        /**
         * An event emitted when the user clicks the search button.
         */
        this.buttonClick = new EventEmitter();
        this._buttonLabel = null;
        this._hideSearchButton = false;
        this._buttonLayout = '12,12,12,2';
        this._contentLayout = null;
    }
    /**
     * Sets the text label of the button.
     * @param {?} value
     * @return {?}
     */
    set buttonLabel(value) {
        if (this._buttonLabel !== value) {
            this._buttonLabel = value;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get buttonLabel() {
        return this._buttonLabel;
    }
    /**
     * Whether the search button should be hidden.
     * @param {?} value
     * @return {?}
     */
    set hideSearchButton(value) {
        this._hideSearchButton = coerceBooleanProperty(value);
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get hideSearchButton() {
        return this._hideSearchButton;
    }
    /**
     * Layout of the search button inside of a NxGrid. Default: 12,12,12,2.
     * @param {?} value
     * @return {?}
     */
    set buttonLayout(value) {
        if (this._buttonLayout !== value) {
            this._buttonLayout = value;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get buttonLayout() {
        return this._buttonLayout;
    }
    /**
     * Layout of the content area inside of a NxGrid. Default: 12,12,12,10 | 12 (if search button is hidden).
     * @param {?} value
     * @return {?}
     */
    set contentLayout(value) {
        if (this._contentLayout !== value) {
            this._contentLayout = value;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get contentLayout() {
        if (this._contentLayout) {
            return this._contentLayout;
        }
        return !this.hideSearchButton ? '12,12,12,10' : '12';
    }
    /**
     * \@docs-private
     * @return {?}
     */
    onButtonClick() {
        this.buttonClick.emit();
    }
}
NxPageSearchComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-page-search',
                template: "<div class=\"nx-pagesearch\">\n  <div nxLayout=\"grid\">\n    <div nxRow>\n      <div [nxCol]=\"contentLayout\" class=\"nx-pagesearch--content\">\n        <ng-content></ng-content>\n      </div>\n      <div [nxCol]=\"buttonLayout\" *ngIf=\"!hideSearchButton\" class=\"nx-pagesearch__actions\">\n        <button nxButton=\"primary medium\" (click)=\"onButtonClick()\">{{buttonLabel}}</button>\n      </div>\n    </div>\n  </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".nx-pagesearch{padding:32px 0;background-color:#fff;box-shadow:0 2px 4px rgba(65,65,65,.5)}.nx-pagesearch ::ng-deep .nx-formfield__wrapper{padding-bottom:0}.nx-pagesearch ::ng-deep .c-input{font-size:30px;font-weight:300;line-height:32px;height:auto}.nx-pagesearch ::ng-deep .nx-formfield__prefix{color:#007ab3}.nx-pagesearch ::ng-deep button{margin:0}.nx-pagesearch ::ng-deep .nx-icon--auto{font-size:24px}.nx-pagesearch .nx-pagesearch__actions{display:flex;align-items:center}"]
            }] }
];
/** @nocollapse */
NxPageSearchComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
NxPageSearchComponent.propDecorators = {
    buttonClick: [{ type: Output, args: ['nxButtonClick',] }],
    buttonLabel: [{ type: Input, args: ['nxButtonLabel',] }],
    hideSearchButton: [{ type: Input, args: ['nxHideSearchButton',] }],
    buttonLayout: [{ type: Input, args: ['nxButtonLayout',] }],
    contentLayout: [{ type: Input, args: ['nxContentLayout',] }]
};
if (false) {
    /**
     * An event emitted when the user clicks the search button.
     * @type {?}
     */
    NxPageSearchComponent.prototype.buttonClick;
    /**
     * @type {?}
     * @private
     */
    NxPageSearchComponent.prototype._buttonLabel;
    /**
     * @type {?}
     * @private
     */
    NxPageSearchComponent.prototype._hideSearchButton;
    /**
     * @type {?}
     * @private
     */
    NxPageSearchComponent.prototype._buttonLayout;
    /**
     * @type {?}
     * @private
     */
    NxPageSearchComponent.prototype._contentLayout;
    /**
     * @type {?}
     * @private
     */
    NxPageSearchComponent.prototype._changeDetectorRef;
}
