// @allianz/ngx-ndbx/esm5/datefield/datefield.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NxIconModule } from '@allianz/ngx-ndbx/icon';
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NX_DATE_LOCALE_PROVIDER } from './adapter/index';
import { NxDatefieldDirective } from './datefield.directive';
import { NxCalendarComponent } from './datepicker/calendar';
import { NxCalendarBodyComponent } from './datepicker/calendar-body';
import { NxDatepickerIntl } from './datepicker/datepicker-intl';
import { NxDatepickerToggleComponent, NxDatepickerToggleIconComponent } from './datepicker/datepicker-toggle';
import { NX_DATEPICKER_SCROLL_STRATEGY_PROVIDER, NxDatepickerComponent, NxDatepickerContentComponent, } from './datepicker/datepicker.component';
import { NxMonthViewComponent } from './datepicker/month-view';
import { NxMultiYearViewComponent } from './datepicker/multi-year-view';
import { NxYearViewComponent } from './datepicker/year-view';
var NxDatefieldModule = /** @class */ (function () {
    function NxDatefieldModule() {
    }
    NxDatefieldModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        OverlayModule,
                        A11yModule,
                        NxIconModule
                    ],
                    providers: [
                        // Provide our NX_DATE_LOCALE to determine the current language
                        NX_DATE_LOCALE_PROVIDER,
                        NxDatepickerIntl,
                        NX_DATEPICKER_SCROLL_STRATEGY_PROVIDER
                    ],
                    declarations: [
                        NxDatefieldDirective,
                        NxDatepickerComponent,
                        NxDatepickerToggleIconComponent,
                        NxDatepickerToggleComponent,
                        NxDatepickerContentComponent,
                        NxCalendarComponent,
                        NxCalendarBodyComponent,
                        NxMonthViewComponent,
                        NxMultiYearViewComponent,
                        NxYearViewComponent
                    ],
                    exports: [
                        NxDatefieldDirective,
                        NxDatepickerComponent,
                        NxDatepickerToggleIconComponent,
                        NxDatepickerToggleComponent,
                        NxCalendarComponent,
                        NxCalendarBodyComponent,
                        NxMonthViewComponent,
                        NxMultiYearViewComponent,
                        NxYearViewComponent
                    ],
                    entryComponents: [
                        NxDatepickerContentComponent
                    ]
                },] }
    ];
    return NxDatefieldModule;
}());
export { NxDatefieldModule };
