// @allianz/ngx-ndbx/esm5/circle-toggle/mobile-toggle-button/mobile-toggle-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
/**
 * \@docs-private
 */
var NxMobileToggleButtonComponent = /** @class */ (function () {
    function NxMobileToggleButtonComponent(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this.checked = false;
        this.negative = false;
        this.disabled = false;
    }
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxMobileToggleButtonComponent.prototype.setFirstButton = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this.isFirst = true;
        this._changeDetectorRef.markForCheck();
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxMobileToggleButtonComponent.prototype.setLastButton = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this.isLast = true;
        this._changeDetectorRef.markForCheck();
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxMobileToggleButtonComponent.prototype.resetClasses = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this.isFirst = false;
        this.isLast = false;
        this._changeDetectorRef.markForCheck();
    };
    NxMobileToggleButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-mobile-toggle-button',
                    template: "<nx-icon class=\"nx-toggle-button__mobile-check\" name=\"check\" aria-hidden=\"true\"></nx-icon>\n<span class=\"nx-toggle-button__label-text\">{{label}}</span>\n<span *ngIf=\"hint\" class=\"nx-toggle-button__label-hint\">({{hint}})</span>\n\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.is-disabled]': 'disabled',
                        '[class.is-negative]': 'negative',
                        '[class.is-first]': 'isFirst',
                        '[class.is-last]': 'isLast',
                        '[class.is-flipped]': 'checked'
                    },
                    styles: [":host{display:none;position:relative;transition:background-color .2s,border-color .2s,color .2s;flex-direction:row;border-width:1px 2px;border-color:#007ab3;border-style:solid;padding:11px 31px;width:100%;justify-content:center;align-items:center}@media (max-width:703px){:host-context(.is-responsive){display:flex}}.nx-toggle-button__label-hint{color:#007ab3;font-weight:nx-font-weights(light);line-height:20px;font-size:14px}.nx-toggle-button__label-text{padding-top:0;padding-right:4px;font-weight:600;color:#007ab3}.nx-toggle-button__mobile-check{font-size:24px;height:24px;display:inline-block;opacity:0;margin-right:8px;color:#fff}:host(.is-flipped.nx-toggle-button__mobile-check){opacity:1}:host(.is-first){border-top-width:2px;border-radius:4px 4px 0 0}:host(.is-last){border-bottom-width:2px;border-radius:0 0 4px 4px}:host(.is-flipped){background:#007ab3}:host(.is-flipped) .nx-toggle-button__label-hint,:host(.is-flipped) .nx-toggle-button__label-text{color:#fff}:host(.is-negative){border-color:#fff}:host(.is-negative) .nx-toggle-button__label-hint,:host(.is-negative) .nx-toggle-button__label-text,:host(.is-negative) .nx-toggle-button__mobile-check{color:#fff}:host(.is-negative.is-flipped){background-color:#fff}:host(.is-negative.is-flipped) .nx-toggle-button__label-hint,:host(.is-negative.is-flipped) .nx-toggle-button__label-text,:host(.is-negative.is-flipped) .nx-toggle-button__mobile-check{color:#007ab3}:host(.is-disabled){opacity:.4;cursor:not-allowed}"]
                }] }
    ];
    /** @nocollapse */
    NxMobileToggleButtonComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NxMobileToggleButtonComponent.propDecorators = {
        checked: [{ type: Input }],
        negative: [{ type: Input }],
        disabled: [{ type: Input }],
        hint: [{ type: Input }],
        label: [{ type: Input }]
    };
    return NxMobileToggleButtonComponent;
}());
export { NxMobileToggleButtonComponent };
if (false) {
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.checked;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.negative;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.disabled;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.hint;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.label;
    /**
     * \@docs-private
     * @type {?}
     */
    NxMobileToggleButtonComponent.prototype.isFirst;
    /**
     * \@docs-private
     * @type {?}
     */
    NxMobileToggleButtonComponent.prototype.isLast;
    /**
     * @type {?}
     * @private
     */
    NxMobileToggleButtonComponent.prototype._changeDetectorRef;
}
