// @allianz/ngx-ndbx/esm2015/popover/popover.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ContentChild, EventEmitter, Output, TemplateRef, ViewChild } from '@angular/core';
import { Subject } from 'rxjs';
import { NxPopoverContentDirective } from './popover-content';
export class NxPopoverComponent {
    constructor() {
        /**
         * Event emitted when the popover is closed.
         */
        this.closed = new EventEmitter();
        /**
         * \@docs-private
         */
        this.closeButtonClick = new Subject();
        /**
         * \@docs-private
         */
        this.showCloseButton = false;
        /**
         * \@docs-private
         */
        this.arrowStyle = {};
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.closed.complete();
    }
    // emit to notify the popover trigger directive that the close button was clicked
    /**
     * \@docs-private
     * @return {?}
     */
    emitButtonClick() {
        this.closeButtonClick.next();
    }
    /**
     * \@docs-private
     * @return {?}
     */
    emitClosedEvent() {
        this.closed.emit();
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get classList() {
        if (this.direction) {
            return [`nx-popover--${this.direction}`];
        }
    }
}
NxPopoverComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-popover',
                template: "<ng-template>\n  <div [ngClass]=\"classList\" role=\"tooltip\" aria-hidden=\"false\">\n    <div class=\"nx-popover__content\">\n        <div class=\"nx-popover__arrow\" [ngStyle]=\"arrowStyle\">\n          </div>\n      <span aria-hidden=\"true\" class=\"nx-popover__close-icon\" (click)=\"emitButtonClick()\"\n            *ngIf=\"showCloseButton\">\n        <nx-icon name=\"close\"></nx-icon>\n      </span>\n      <span [id]=\"id\">\n        <ng-content></ng-content>\n        <ng-container [ngTemplateOutlet]=\"_lazyContent?._template\"></ng-container>\n      </span>\n    </div>\n  </div>\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                exportAs: 'nxPopover',
                styles: [".nx-popover__content{background-color:#fff;border:1px solid #d9d9d9;border-radius:4px;box-shadow:0 2px 4px rgba(65,65,65,.5);padding:23px;display:inline-block;font-size:16px}.nx-popover__close-icon{position:absolute;top:4px;right:4px;cursor:pointer;font-size:16px;line-height:16px}[data-whatintent=mouse] .nx-popover__content:focus{outline:0}.nx-popover__arrow{width:12px;height:12px;background-color:inherit;position:absolute;-webkit-transform:translate(-50%,0);transform:translate(-50%,0);border-top:1px solid #d9d9d9;border-left:1px solid #d9d9d9}.nx-popover--top .nx-popover__arrow{-webkit-transform:translate(-50%,0) rotate(225deg);transform:translate(-50%,0) rotate(225deg);bottom:-5px;box-shadow:2px 2px 1px 1px #fff,-2px -2px 4px rgba(65,65,65,.5)}.nx-popover--right .nx-popover__arrow{-webkit-transform:translate(0,-50%) rotate(-45deg);transform:translate(0,-50%) rotate(-45deg);left:-5px}.nx-popover--bottom .nx-popover__arrow{-webkit-transform:translate(-50%,0) rotate(45deg);transform:translate(-50%,0) rotate(45deg);top:-5px}.nx-popover--left .nx-popover__arrow{-webkit-transform:translate(0,-50%) rotate(135deg);transform:translate(0,-50%) rotate(135deg);right:-5px}"]
            }] }
];
/** @nocollapse */
NxPopoverComponent.ctorParameters = () => [];
NxPopoverComponent.propDecorators = {
    templateRef: [{ type: ViewChild, args: [TemplateRef,] }],
    _lazyContent: [{ type: ContentChild, args: [NxPopoverContentDirective,] }],
    closed: [{ type: Output, args: ['nxClosed',] }]
};
if (false) {
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.templateRef;
    /**
     * Content that will be rendered lazily.
     * @type {?}
     */
    NxPopoverComponent.prototype._lazyContent;
    /**
     * Event emitted when the popover is closed.
     * @type {?}
     */
    NxPopoverComponent.prototype.closed;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.closeButtonClick;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.id;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.direction;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.showCloseButton;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.arrowStyle;
}
