// @allianz/ngx-ndbx/esm2015/modal/modal.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NxIconModule } from '@allianz/ngx-ndbx/icon';
import { NgModule } from '@angular/core';
import { NxModalComponent, NxModalActionsDirective, NxModalContentDirective } from './modal.component';
import { CommonModule } from '@angular/common';
import { NxOpenModalOnClickDirective } from './modal-open-on-click.directive';
import { NxModalService } from './modal.service';
import { A11yModule } from '@angular/cdk/a11y';
export class NxModalModule {
    // define the modal service on root level as an application wide singleton,
    // assuming that we only display one modal at a time, which seems quite reasonable
    // this is static and named "forRoot" by convention
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: NxModalModule,
            providers: [NxModalService]
        };
    }
}
NxModalModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, A11yModule, NxIconModule],
                declarations: [
                    NxModalComponent,
                    NxOpenModalOnClickDirective,
                    NxModalActionsDirective,
                    NxModalContentDirective
                ],
                exports: [
                    NxModalComponent,
                    NxOpenModalOnClickDirective,
                    NxModalActionsDirective,
                    NxModalContentDirective
                ],
            },] }
];
