// @allianz/ngx-ndbx/esm5/progress-stepper/multi/multi-step-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { coerceBooleanProperty, coerceNumberProperty } from '@angular/cdk/coercion';
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef, forwardRef, Inject } from '@angular/core';
import { NxStepComponent, NxProgressStepperDirective } from './../progress-stepper.component';
import { ErrorStateMatcher } from '@allianz/ngx-ndbx/utils';
/**
 * \@docs-private
 */
var NxMultiStepItemComponent = /** @class */ (function (_super) {
    tslib_1.__extends(NxMultiStepItemComponent, _super);
    function NxMultiStepItemComponent(_changeDetectorRef, stepper, _errorStateMatcher) {
        var _this = _super.call(this, stepper, _errorStateMatcher) || this;
        _this._changeDetectorRef = _changeDetectorRef;
        _this._completed = false;
        return _this;
    }
    Object.defineProperty(NxMultiStepItemComponent.prototype, "index", {
        /** Sets the index of the selected step. */
        get: /**
         * Sets the index of the selected step.
         * @return {?}
         */
        function () { return this._index; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this._index = coerceNumberProperty(value); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "selected", {
        /** Sets the selected step. */
        get: /**
         * Sets the selected step.
         * @return {?}
         */
        function () { return this._selected; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this._selected = coerceBooleanProperty(value); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "active", {
        /** Sets the active step. */
        get: /**
         * Sets the active step.
         * @return {?}
         */
        function () { return this._active; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this._active = coerceBooleanProperty(value); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "last", {
        /** Sets the last step. */
        get: /**
         * Sets the last step.
         * @return {?}
         */
        function () { return this._last; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this._last = coerceBooleanProperty(value); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "first", {
        /** Sets the first step. */
        get: /**
         * Sets the first step.
         * @return {?}
         */
        function () { return this._first; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this._first = coerceBooleanProperty(value); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "stepCompleted", {
        get: /**
         * @return {?}
         */
        function () { return this._completed; },
        /** Whether the step is completed or not. */
        set: /**
         * Whether the step is completed or not.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._completed = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    NxMultiStepItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-multi-step-item',
                    template: "<div class=\"label\">\n    {{label}}\n  </div>\n\n<div\n  [ngSwitch]=\"state\"\n  class=\"dot\"\n  [class.dot--filled]='stepCompleted'>\n  <nx-icon *ngIf=\"stepCompleted\" name=\"check\"></nx-icon>\n  <div *ngIf=\"selected && !stepCompleted\" class=\"small-dot\"></div>\n\n</div>\n<div\n  *ngIf=\"!last\"\n  class=\"before-line\"\n  [class.before-line--active]='active'>\n</div>\n<div\n  *ngIf=\"!last\"\n  class=\"line\"\n  [class.line--active]='active'>\n</div>\n<div\n  *ngIf=\"!last\"\n  class=\"after-line\"\n  [class.after-line--active]='active'>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host{flex:1 0 auto;display:flex;align-items:center;position:relative;font-size:16px;padding-bottom:32px;cursor:pointer}.label{position:absolute;top:calc(8px + 16px);-webkit-transform:translateX(-50%) translateX(8px);transform:translateX(-50%) translateX(8px);min-width:156px;text-align:center;line-height:24px;font-size:16px;color:#414141}.dot{width:16px;height:16px;border:2px solid #007ab3;border-radius:50%;color:#fff;background-color:#fff;flex:0 0 auto}.dot nx-icon{font-size:16px;position:absolute;top:0;left:0}.small-dot{top:4px;left:4px;position:absolute;width:8px;height:8px;border-radius:50%;background-color:#007ab3}.dot--filled{background-color:#007ab3;color:#fff}.line{flex:1 1 100%;min-width:32px;height:0;border-top:4px dotted #c2c2c2}.line--active{border-top-color:#007ab3;border-top-style:solid;border-image-source:none;transition:border-style .1s ease-in}.after-line,.before-line{width:4px;height:4px;z-index:-1}.after-line--active,.before-line--active{z-index:0;background:#007ab3}"]
                }] }
    ];
    /** @nocollapse */
    NxMultiStepItemComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: NxProgressStepperDirective, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return NxProgressStepperDirective; })),] }] },
        { type: ErrorStateMatcher }
    ]; };
    NxMultiStepItemComponent.propDecorators = {
        state: [{ type: Input }],
        label: [{ type: Input }],
        index: [{ type: Input }],
        selected: [{ type: Input }],
        active: [{ type: Input }],
        last: [{ type: Input }],
        first: [{ type: Input }],
        stepCompleted: [{ type: Input }]
    };
    return NxMultiStepItemComponent;
}(NxStepComponent));
export { NxMultiStepItemComponent };
if (false) {
    /**
     * The state of the step.
     * @type {?}
     */
    NxMultiStepItemComponent.prototype.state;
    /**
     * The label of the step.
     * @type {?}
     */
    NxMultiStepItemComponent.prototype.label;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._index;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._selected;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._active;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._last;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._first;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._completed;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._changeDetectorRef;
}
