// @allianz/ngx-ndbx/esm2015/radio-circle/radio-circle-button.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef, Input, ViewChild } from '@angular/core';
/** @type {?} */
let nextId = 0;
/**
 * @abstract
 */
export class NxRadioCircleButton {
    /**
     * @param {?} radioCircle
     */
    constructor(radioCircle) {
        this.radioCircle = radioCircle;
        /**
         * \@docs-private
         */
        this.id = ++nextId;
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        if (!this.label) {
            console.warn(`nx-radio-circle-icon-button/nx-radio-circle-svg-button: required field 'nxLabel' is not set`);
        }
    }
    /**
     * @return {?}
     */
    get checked() {
        return this.radioCircle.selectedButton === this || null;
    }
    /**
     * Whether the item is preselected.
     * @param {?} checked
     * @return {?}
     */
    set checked(checked) {
        if (checked) {
            this.radioCircle.selectedButton = this;
        }
        else {
            if (this.checked) {
                this.radioCircle.selectedButton = null;
            }
        }
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get circleButtonId() {
        return this.radioCircle.radioCircleId + '-' + this.id;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get radioCircleId() {
        return this.radioCircle.radioCircleId;
    }
}
NxRadioCircleButton.propDecorators = {
    label: [{ type: Input, args: ['nxLabel',] }],
    hint: [{ type: Input, args: ['nxHint',] }],
    value: [{ type: Input, args: ['nxValue',] }],
    labelElement: [{ type: ViewChild, args: ['labelElement',] }],
    checked: [{ type: Input, args: ['nxSelected',] }]
};
if (false) {
    /**
     * \@docs-private
     * @type {?}
     */
    NxRadioCircleButton.prototype.id;
    /**
     * Sets the label displayed below the circle.
     * @type {?}
     */
    NxRadioCircleButton.prototype.label;
    /**
     * Sets an additional hint displayed below the label.
     * @type {?}
     */
    NxRadioCircleButton.prototype.hint;
    /**
     * Sets the value used in the model.
     * @type {?}
     */
    NxRadioCircleButton.prototype.value;
    /**
     * \@docs-private
     * @type {?}
     */
    NxRadioCircleButton.prototype.labelElement;
    /**
     * \@docs-private
     * @type {?}
     * @protected
     */
    NxRadioCircleButton.prototype.radioCircle;
}
