// @allianz/ngx-ndbx/esm2015/config/nx-expert.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FORMFIELD_DEFAULT_OPTIONS } from '@allianz/ngx-ndbx/formfield';
import { LABEL_DEFAULT_OPTIONS, ERROR_DEFAULT_OPTIONS } from '@allianz/ngx-ndbx/base';
import { DATEPICKER_DEFAULT_OPTIONS } from '@allianz/ngx-ndbx/datefield';
// expert presets
/** @type {?} */
const formfieldExpertOptions = {
    appearance: 'outline',
    nxFloatLabel: 'always'
};
/** @type {?} */
const errorExpertOptions = { appearance: 'text' };
/** @type {?} */
const labelExpertOptions = { size: 'small' };
/** @type {?} */
const datepickerExpertOptions = { toggleIconTabindex: -1 };
const ɵ0 = formfieldExpertOptions, ɵ1 = labelExpertOptions, ɵ2 = datepickerExpertOptions, ɵ3 = errorExpertOptions;
// should not be a singleton since different tokens can be used in seperate modules
export class NxExpertModule {
}
NxExpertModule.decorators = [
    { type: NgModule, args: [{
                providers: [
                    { provide: FORMFIELD_DEFAULT_OPTIONS, useValue: ɵ0 },
                    { provide: LABEL_DEFAULT_OPTIONS, useValue: ɵ1 },
                    { provide: DATEPICKER_DEFAULT_OPTIONS, useValue: ɵ2 },
                    { provide: ERROR_DEFAULT_OPTIONS, useValue: ɵ3 }
                ]
            },] }
];
export { ɵ0, ɵ1, ɵ2, ɵ3 };
