// @allianz/ngx-ndbx/esm2015/datefield/adapter/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NX_NATIVE_DATE_FORMATS } from './native-date-formats';
import { NxNativeDateAdapter } from './native-date-adapter';
import { NX_DATE_FORMATS } from './date-formats';
import { NgModule } from '@angular/core';
import { LOCALE_ID } from '@angular/core';
import { NxDateAdapter } from './date-adapter';
import { NX_DATE_LOCALE } from './date-token';
import { PlatformModule } from '@angular/cdk/platform';
/** @type {?} */
export const NX_DATE_LOCALE_PROVIDER = { provide: NX_DATE_LOCALE, useExisting: LOCALE_ID };
export { NX_DATE_LOCALE } from './date-token';
export { NxDateAdapter } from './date-adapter';
export { NX_DATE_FORMATS } from './date-formats';
export { NxNativeDateAdapter } from './native-date-adapter';
export { NX_NATIVE_DATE_FORMATS } from './native-date-formats';
export class NativeDateModule {
}
NativeDateModule.decorators = [
    { type: NgModule, args: [{
                imports: [PlatformModule],
                providers: [
                    { provide: NxDateAdapter, useClass: NxNativeDateAdapter },
                    NX_DATE_LOCALE_PROVIDER
                ],
            },] }
];
const ɵ0 = NX_NATIVE_DATE_FORMATS;
export class NxNativeDateModule {
}
NxNativeDateModule.decorators = [
    { type: NgModule, args: [{
                imports: [NativeDateModule],
                providers: [{ provide: NX_DATE_FORMATS, useValue: ɵ0 }],
            },] }
];
export { ɵ0 };
