// @allianz/ngx-ndbx/fesm2015/allianz-ngx-ndbx-footer.js
import { Component, ChangeDetectionStrategy, Directive, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NxFooterCopyrightDirective {
}
NxFooterCopyrightDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'nx-footer-copyright',
                exportAs: 'NxFooterCopyright',
                host: {
                    'class': 'nx-footer__copyright'
                }
            },] }
];
class NxFooterNavigationDirective {
}
NxFooterNavigationDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'nx-footer-navigation',
                exportAs: 'NxFooterNavigation',
                host: {
                    'class': 'nx-footer__navigation',
                    'role': 'list'
                }
            },] }
];
class NxFooterLinkDirective {
}
NxFooterLinkDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'nx-footer-link',
                exportAs: 'NxFooterLink',
                host: {
                    'class': 'nx-footer__link',
                    'role': 'listitem'
                }
            },] }
];
class NxFooterComponent {
}
NxFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-footer, [nx-footer]',
                template: '<ng-content></ng-content>',
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    'class': 'nx-footer'
                },
                styles: [":host(.nx-footer){display:flex;flex-direction:row;color:#767676;background:0 0;padding:72px 32px 24px;font-size:16px}::ng-deep a{text-decoration:none}::ng-deep .nx-footer__copyright{align-self:flex-end;font-weight:700;font-size:inherit;line-height:inherit;color:inherit}@media (max-width:703px){:host(.nx-footer){flex-direction:column-reverse;padding:72px 16px 40px}::ng-deep .nx-footer__copyright{-ms-grid-row-align:center;align-self:center;text-align:center}}::ng-deep .nx-footer__navigation{display:flex;align-self:flex-end}@media (max-width:703px){::ng-deep .nx-footer__navigation{text-align:center;-ms-grid-row-align:center;align-self:center;flex-direction:column}::ng-deep .nx-footer__navigation .nx-footer__link{padding-left:0;padding-bottom:16px}}::ng-deep .nx-footer__link{font-size:inherit;color:inherit;line-height:inherit;font-weight:400;padding-left:32px}:host-context([data-whatinput=keyboard]) ::ng-deep a:focus{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;border-radius:4px}"]
            }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NxFooterModule {
}
NxFooterModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [NxFooterComponent, NxFooterCopyrightDirective, NxFooterLinkDirective, NxFooterNavigationDirective],
                exports: [NxFooterComponent, NxFooterCopyrightDirective, NxFooterLinkDirective, NxFooterNavigationDirective]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { NxFooterModule, NxFooterComponent as ɵd, NxFooterCopyrightDirective as ɵa, NxFooterLinkDirective as ɵc, NxFooterNavigationDirective as ɵb };
