// @allianz/ngx-ndbx/esm5/progress-stepper/buttons.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CdkStepper, CdkStepperNext, CdkStepperPrevious } from '@angular/cdk/stepper';
import { ChangeDetectorRef, Directive } from '@angular/core';
import { NxProgressStepperDirective } from './progress-stepper.component';
// tslint:disable:use-input-property-decorator
var NxStepperNextDirective = /** @class */ (function (_super) {
    tslib_1.__extends(NxStepperNextDirective, _super);
    function NxStepperNextDirective(_stepper, changeDetectorRef) {
        var _this = _super.call(this, _stepper) || this;
        _this.changeDetectorRef = changeDetectorRef;
        /**
         * \@docs-private
         */
        _this.disabled = false;
        return _this;
    }
    /**
     * Implemented to prevent changed after checked error after stepper init.
     * When the stepper initializes the directive is checked first before
     * the stepper can check its content thus it doesn't know about it's children yet
     * so _stepper.hasNext returns false and disables the buttons. In the same CD cycle
     * after the stepper has checked its contents the button needs to be enabled resulting
     * in the error.
     */
    /**
     * Implemented to prevent changed after checked error after stepper init.
     * When the stepper initializes the directive is checked first before
     * the stepper can check its content thus it doesn't know about it's children yet
     * so _stepper.hasNext returns false and disables the buttons. In the same CD cycle
     * after the stepper has checked its contents the button needs to be enabled resulting
     * in the error.
     * @return {?}
     */
    NxStepperNextDirective.prototype.ngDoCheck = /**
     * Implemented to prevent changed after checked error after stepper init.
     * When the stepper initializes the directive is checked first before
     * the stepper can check its content thus it doesn't know about it's children yet
     * so _stepper.hasNext returns false and disables the buttons. In the same CD cycle
     * after the stepper has checked its contents the button needs to be enabled resulting
     * in the error.
     * @return {?}
     */
    function () {
        /** @type {?} */
        var stepper = (/** @type {?} */ (this._stepper));
        if (this.disabled !== !stepper.hasNext) {
            this.disabled = !stepper.hasNext;
            this.changeDetectorRef.markForCheck();
        }
    };
    NxStepperNextDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'button[nxStepperNext]',
                    host: {
                        '(click)': '_stepper.next()',
                        '[type]': 'type',
                        '[disabled]': 'disabled'
                    },
                    inputs: ['type'],
                    providers: [{ provide: CdkStepper, useExisting: NxProgressStepperDirective }]
                },] }
    ];
    /** @nocollapse */
    NxStepperNextDirective.ctorParameters = function () { return [
        { type: CdkStepper },
        { type: ChangeDetectorRef }
    ]; };
    return NxStepperNextDirective;
}(CdkStepperNext));
export { NxStepperNextDirective };
if (false) {
    /**
     * \@docs-private
     * @type {?}
     */
    NxStepperNextDirective.prototype.disabled;
    /**
     * @type {?}
     * @private
     */
    NxStepperNextDirective.prototype.changeDetectorRef;
}
var NxStepperPreviousDirective = /** @class */ (function (_super) {
    tslib_1.__extends(NxStepperPreviousDirective, _super);
    function NxStepperPreviousDirective(_stepper, changeDetectorRef) {
        var _this = _super.call(this, _stepper) || this;
        _this.changeDetectorRef = changeDetectorRef;
        return _this;
    }
    NxStepperPreviousDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'button[nxStepperPrevious]',
                    host: {
                        '(click)': '_stepper.previous()',
                        '[type]': 'type',
                        '[disabled]': '!_stepper.hasPrevious'
                    },
                    inputs: ['type'],
                    providers: [{ provide: CdkStepper, useExisting: NxProgressStepperDirective }]
                },] }
    ];
    /** @nocollapse */
    NxStepperPreviousDirective.ctorParameters = function () { return [
        { type: CdkStepper },
        { type: ChangeDetectorRef }
    ]; };
    return NxStepperPreviousDirective;
}(CdkStepperPrevious));
export { NxStepperPreviousDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxStepperPreviousDirective.prototype.changeDetectorRef;
}
