// @allianz/ngx-ndbx/esm2015/button/button-base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DEFAULT_SIZE = 'medium';
/** @type {?} */
const DEFAULT_TYPE = 'primary';
/** @type {?} */
export const NX_BUTTON_HOST_BINDINGS = {
    '[class.nx-button--primary]': 'type === "primary"',
    '[class.nx-button--secondary]': 'type === "secondary"',
    '[class.nx-button--tertiary]': 'type === "tertiary"',
    '[class.nx-button--cta]': 'type === "cta"',
    '[class.nx-button--emphasis]': 'type === "emphasis"',
    '[class.nx-button--large]': 'size === "large"',
    '[class.nx-button--medium]': 'size === "medium"',
    '[class.nx-button--small-medium]': 'size === "small-medium"',
    '[class.nx-button--small]': 'size === "small"',
    '[class.nx-button--block]': 'block',
    '[class.nx-button--negative]': 'negative',
};
/**
 * \@docs-private
 */
export class NxButtonBase {
    /**
     * @param {?} _changeDetectorRef
     * @param {?} _elementRef
     */
    constructor(_changeDetectorRef, _elementRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._elementRef = _elementRef;
        /**
         * \@docs-private
         */
        this.type = DEFAULT_TYPE;
        /**
         * \@docs-private
         */
        this.size = DEFAULT_SIZE;
        /**
         * \@docs-private
         */
        this.negative = false;
        /**
         * \@docs-private
         */
        this.block = false;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set classNames(value) {
        if (this._classNames === value) {
            return;
        }
        this._classNames = value;
        const [type = null] = this._classNames.match(/primary|secondary|tertiary|cta|emphasis/) || [DEFAULT_TYPE];
        this.type = (/** @type {?} */ (type));
        const [size = null] = this._classNames.match(/small-medium|small|medium|large/) || [DEFAULT_SIZE];
        this.size = (/** @type {?} */ (size));
        this.negative = /negative/.test(this._classNames);
        this.block = /block/.test(this._classNames);
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get classNames() {
        return this._classNames;
    }
    /**
     * \@docs-private
     * getter used for the modal component as a quickfix
     * since button got changed from directive to component the reference used by the ngOpenModelOnClick directive
     * for nxButtons is a reference to component instance instead of an element reference. As a workaround we need a
     * way to reach the elementRef of the component until the modal gets refactored.
     * @return {?}
     */
    get elementRef() {
        return this._elementRef;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxButtonBase.prototype._classNames;
    /**
     * \@docs-private
     * @type {?}
     */
    NxButtonBase.prototype.type;
    /**
     * \@docs-private
     * @type {?}
     */
    NxButtonBase.prototype.size;
    /**
     * \@docs-private
     * @type {?}
     */
    NxButtonBase.prototype.negative;
    /**
     * \@docs-private
     * @type {?}
     */
    NxButtonBase.prototype.block;
    /**
     * @type {?}
     * @private
     */
    NxButtonBase.prototype._changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    NxButtonBase.prototype._elementRef;
}
