// @allianz/ngx-ndbx/bundles/allianz-ngx-ndbx-footer.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@allianz/ngx-ndbx/footer', ['exports', '@angular/core'], factory) :
    (factory((global.allianz = global.allianz || {}, global.allianz['ngx-ndbx'] = global.allianz['ngx-ndbx'] || {}, global.allianz['ngx-ndbx'].footer = {}),global.ng.core));
}(this, (function (exports,core) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NxFooterCopyrightDirective = /** @class */ (function () {
        function NxFooterCopyrightDirective() {
        }
        NxFooterCopyrightDirective.decorators = [
            { type: core.Directive, args: [{
                        // tslint:disable-next-line:directive-selector
                        selector: 'nx-footer-copyright',
                        exportAs: 'NxFooterCopyright',
                        host: {
                            'class': 'nx-footer__copyright'
                        }
                    },] }
        ];
        return NxFooterCopyrightDirective;
    }());
    var NxFooterNavigationDirective = /** @class */ (function () {
        function NxFooterNavigationDirective() {
        }
        NxFooterNavigationDirective.decorators = [
            { type: core.Directive, args: [{
                        // tslint:disable-next-line:directive-selector
                        selector: 'nx-footer-navigation',
                        exportAs: 'NxFooterNavigation',
                        host: {
                            'class': 'nx-footer__navigation',
                            'role': 'list'
                        }
                    },] }
        ];
        return NxFooterNavigationDirective;
    }());
    var NxFooterLinkDirective = /** @class */ (function () {
        function NxFooterLinkDirective() {
        }
        NxFooterLinkDirective.decorators = [
            { type: core.Directive, args: [{
                        // tslint:disable-next-line:directive-selector
                        selector: 'nx-footer-link',
                        exportAs: 'NxFooterLink',
                        host: {
                            'class': 'nx-footer__link',
                            'role': 'listitem'
                        }
                    },] }
        ];
        return NxFooterLinkDirective;
    }());
    var NxFooterComponent = /** @class */ (function () {
        function NxFooterComponent() {
        }
        NxFooterComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'nx-footer, [nx-footer]',
                        template: '<ng-content></ng-content>',
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        host: {
                            'class': 'nx-footer'
                        },
                        styles: [":host(.nx-footer){display:flex;flex-direction:row;color:#767676;background:0 0;padding:72px 32px 24px;font-size:16px}::ng-deep a{text-decoration:none}::ng-deep .nx-footer__copyright{align-self:flex-end;font-weight:700;font-size:inherit;line-height:inherit;color:inherit}@media (max-width:703px){:host(.nx-footer){flex-direction:column-reverse;padding:72px 16px 40px}::ng-deep .nx-footer__copyright{-ms-grid-row-align:center;align-self:center;text-align:center}}::ng-deep .nx-footer__navigation{display:flex;align-self:flex-end}@media (max-width:703px){::ng-deep .nx-footer__navigation{text-align:center;-ms-grid-row-align:center;align-self:center;flex-direction:column}::ng-deep .nx-footer__navigation .nx-footer__link{padding-left:0;padding-bottom:16px}}::ng-deep .nx-footer__link{font-size:inherit;color:inherit;line-height:inherit;font-weight:400;padding-left:32px}:host-context([data-whatinput=keyboard]) ::ng-deep a:focus{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;border-radius:4px}"]
                    }] }
        ];
        return NxFooterComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NxFooterModule = /** @class */ (function () {
        function NxFooterModule() {
        }
        NxFooterModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        declarations: [NxFooterComponent, NxFooterCopyrightDirective, NxFooterLinkDirective, NxFooterNavigationDirective],
                        exports: [NxFooterComponent, NxFooterCopyrightDirective, NxFooterLinkDirective, NxFooterNavigationDirective]
                    },] }
        ];
        return NxFooterModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NxFooterModule = NxFooterModule;
    exports.ɵd = NxFooterComponent;
    exports.ɵa = NxFooterCopyrightDirective;
    exports.ɵc = NxFooterLinkDirective;
    exports.ɵb = NxFooterNavigationDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
