// @allianz/ngx-ndbx/bundles/allianz-ngx-ndbx-progressbar.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@allianz/ngx-ndbx/utils')) :
    typeof define === 'function' && define.amd ? define('@allianz/ngx-ndbx/progressbar', ['exports', '@angular/common', '@angular/core', '@allianz/ngx-ndbx/utils'], factory) :
    (factory((global.allianz = global.allianz || {}, global.allianz['ngx-ndbx'] = global.allianz['ngx-ndbx'] || {}, global.allianz['ngx-ndbx'].progressbar = {}),global.ng.common,global.ng.core,global.allianz['ngx-ndbx'].utils));
}(this, (function (exports,common,core,utils) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var progressbarId = 0;
    var NxProgressbarComponent = /** @class */ (function () {
        function NxProgressbarComponent() {
            /**
             * \@docs-private
             */
            this.progressbarId = "nx-progress-bar-" + progressbarId++;
            this._value = 0;
        }
        Object.defineProperty(NxProgressbarComponent.prototype, "value", {
            /** Sets the value of the progress bar. Defaults to zero. Mirrored to aria-value now. */
            get: /**
             * Sets the value of the progress bar. Defaults to zero. Mirrored to aria-value now.
             * @return {?}
             */ function () { return this._value; },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) { this._value = utils.clamp(value || 0); },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        NxProgressbarComponent.prototype._primaryTransform = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var scale = this.value;
                return { transform: "scaleX(" + scale + ")" };
            };
        NxProgressbarComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'nx-progressbar',
                        template: "<div class=\"nx-progress__track\">\n  <div class=\"nx-progress__indicator\" [ngStyle]=\"_primaryTransform()\"></div>\n</div>",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        host: {
                            '[attr.aria-valuenow]': 'value'
                        },
                        styles: [":host{display:block;height:4px;overflow:hidden;position:relative;transition:opacity 250ms linear;width:100%}.nx-progress__track{background-color:#ececec;height:100%}.nx-progress__indicator{-webkit-transform-origin:top left;transform-origin:top left;height:100%;position:absolute;width:100%}.nx-progress__indicator::after{height:100%;position:absolute;width:100%;-webkit-animation:none;animation:none;content:'';display:inline-block;left:0;background:#007ab3}"]
                    }] }
        ];
        NxProgressbarComponent.propDecorators = {
            value: [{ type: core.Input }]
        };
        return NxProgressbarComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NxProgressbarModule = /** @class */ (function () {
        function NxProgressbarModule() {
        }
        NxProgressbarModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        exports: [NxProgressbarComponent],
                        declarations: [NxProgressbarComponent],
                        providers: [],
                    },] }
        ];
        return NxProgressbarModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NxProgressbarModule = NxProgressbarModule;
    exports.NxProgressbarComponent = NxProgressbarComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
