// @allianz/ngx-ndbx/bundles/allianz-ngx-ndbx-core.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@allianz/ngx-ndbx/utils')) :
    typeof define === 'function' && define.amd ? define('@allianz/ngx-ndbx/core', ['exports', '@allianz/ngx-ndbx/utils'], factory) :
    (factory((global.allianz = global.allianz || {}, global.allianz['ngx-ndbx'] = global.allianz['ngx-ndbx'] || {}, global.allianz['ngx-ndbx'].core = {}),global.allianz['ngx-ndbx'].utils));
}(this, (function (exports,utils) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var BREAKPOINT_MOBILE = '(max-width: 703px)';
    /** @type {?} */
    var BREAKPOINT_TABLET = '(min-width: 704px) and (max-width: 992px)';
    /** @type {?} */
    var BREAKPOINT_DESKTOP = '(min-width: 993px)';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MappedStyles = /** @class */ (function () {
        function MappedStyles(mapping, baseClasses, _elementRef, _renderer) {
            if (baseClasses === void 0) {
                baseClasses = [];
            }
            this.mapping = mapping;
            this.baseClasses = baseClasses;
            this._elementRef = _elementRef;
            this._renderer = _renderer;
            this.classNames = '';
        }
        Object.defineProperty(MappedStyles.prototype, "classNames", {
            get: /**
             * @return {?}
             */ function () {
                return this._classNames;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                if (this._classNames === value) {
                    return;
                }
                utils.removeClasses(this._renderer, this._elementRef, this._classNamesSanitized);
                this._classNamesSanitized = utils.mapClassNames(value, this.baseClasses, this.mapping);
                this._classNames = value;
                utils.appendClasses(this._renderer, this._elementRef, this._classNamesSanitized);
            },
            enumerable: true,
            configurable: true
        });
        return MappedStyles;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.BREAKPOINT_MOBILE = BREAKPOINT_MOBILE;
    exports.BREAKPOINT_TABLET = BREAKPOINT_TABLET;
    exports.BREAKPOINT_DESKTOP = BREAKPOINT_DESKTOP;
    exports.MappedStyles = MappedStyles;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
