// @allianz/ngx-ndbx/bundles/allianz-ngx-ndbx-image.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@allianz/ngx-ndbx/image', ['exports', '@angular/core'], factory) :
    (factory((global.allianz = global.allianz || {}, global.allianz['ngx-ndbx'] = global.allianz['ngx-ndbx'] || {}, global.allianz['ngx-ndbx'].image = {}),global.ng.core));
}(this, (function (exports,core) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NxImageDirective = /** @class */ (function () {
        function NxImageDirective() {
            console.warn('Directive nxImg which was used together with nxFigure ' +
                'is now deprecated and not required anymore. Please remove it from your ' +
                'tag. Check the documentation for more details.');
        }
        NxImageDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'img[nxImg]'
                    },] }
        ];
        /** @nocollapse */
        NxImageDirective.ctorParameters = function () { return []; };
        return NxImageDirective;
    }());

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var DEFAULT_SIZE = 'auto';
    var NxFigureComponent = /** @class */ (function () {
        function NxFigureComponent() {
        }
        Object.defineProperty(NxFigureComponent.prototype, "classNames", {
            get: /**
             * @return {?}
             */ function () {
                return this._classNames;
            },
            /**
             * Sets the type of the visual appearance of the image.
             * The default value  is 'auto'.
             */
            set: /**
             * Sets the type of the visual appearance of the image.
             * The default value  is 'auto'.
             * @param {?} value
             * @return {?}
             */ function (value) {
                if (this._classNames === value) {
                    return;
                }
                this._classNames = value;
                /** @type {?} */
                var sizeRegex = /^(auto|1by1|1dot8by1|1dot2by1|1by1dot1|2dot6by1)$/;
                var _a = __read(this._classNames.match(sizeRegex) || [DEFAULT_SIZE], 1), _b = _a[0], size = _b === void 0 ? null : _b;
                this.size = ( /** @type {?} */(size));
                this.rounded = !!this._classNames.match(/rounded/);
            },
            enumerable: true,
            configurable: true
        });
        NxFigureComponent.decorators = [
            { type: core.Component, args: [{
                        // tslint:disable-next-line:component-selector
                        selector: 'figure[nxFigure]',
                        template: '<ng-content></ng-content>',
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        host: {
                            '[class.nx-image--auto]': 'size === "auto" && !rounded',
                            '[class.nx-image--1by1]': 'size === "1by1"',
                            '[class.nx-image--1dot8by1]': 'size === "1dot8by1"',
                            '[class.nx-image--1dot2by1]': 'size === "1dot2by1"',
                            '[class.nx-image--1by1dot1]': 'size === "1by1dot1"',
                            '[class.nx-image--2dot6by1]': 'size === "2dot6by1"',
                            '[class.nx-image--rounded]': 'rounded'
                        },
                        styles: [":host{display:block;position:relative;line-height:0}:host ::ng-deep img{display:block;font-family:\"object-fit: cover;\";-o-object-fit:cover;object-fit:cover;height:100%;width:100%}:host(.nx-image--auto) ::ng-deep img{height:auto}:host(:not(.nx-image--auto)) ::ng-deep img{position:absolute;top:0;left:0}:host(.nx-image--rounded){padding-top:100%}:host(.nx-image--rounded) ::ng-deep img{border-radius:50%;bottom:0;left:0;position:absolute;right:0;top:0}:host(.nx-image--1by1){padding-top:100%}:host(.nx-image--1dot8by1){padding-top:55.55556%}:host(.nx-image--1dot2by1){padding-top:83.33333%}:host(.nx-image--1by1dot1){padding-top:110%}:host(.nx-image--2dot6by1){padding-top:38.46154%}:host(.nx-image--1by1dot4){padding-top:140%}"]
                    }] }
        ];
        NxFigureComponent.propDecorators = {
            classNames: [{ type: core.Input, args: ['nxFigure',] }]
        };
        return NxFigureComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NxImageModule = /** @class */ (function () {
        function NxImageModule() {
        }
        NxImageModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [NxFigureComponent, NxImageDirective],
                        exports: [NxFigureComponent, NxImageDirective]
                    },] }
        ];
        return NxImageModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NxImageModule = NxImageModule;
    exports.NxFigureComponent = NxFigureComponent;
    exports.NxImageDirective = NxImageDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
