// @allianz/ngx-ndbx/esm5/grid/utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mapClassNames } from '@allianz/ngx-ndbx/utils';
/**
 * is udefined or empty?
 * @param {?} value
 * @return {?}
 */
export function isEmpty(value) {
    return (value === undefined || value === '');
}
/**
 * @param {?} value
 * @return {?}
 */
export function isEmptyArray(value) {
    return (value === undefined || value.length === 0);
}
/**
 * add space and return mapClassNames (css)
 * @param {?} aux
 * @param {?} map
 * @return {?}
 */
export function addStyles(aux, map) {
    return ' ' + mapClassNames(aux, [], map);
}
/**
 * input='small,medium,large'
 *  map = MAPPING
 * @param {?} input
 * @param {?} map
 * @return {?}
 */
export function addStylesFromDimensions(input, map) {
    /** @type {?} */
    var aux = processSplit(input);
    /** @type {?} */
    var output = '';
    if (aux.length >= 1 && aux.length <= 4) {
        /** @type {?} */
        var auxTiny = addStyles(aux[0], map).replace('-{tier}', '');
        /** @type {?} */
        var auxSmall = '';
        /** @type {?} */
        var auxMedium = '';
        /** @type {?} */
        var auxLarge = '';
        switch (aux.length) {
            case 1:
                auxSmall = addStyles(aux[0], map).replace('{tier}', 'small');
                auxMedium = addStyles(aux[0], map).replace('{tier}', 'medium');
                auxLarge = addStyles(aux[0], map).replace('{tier}', 'large');
                break;
            case 2:
                auxSmall = addStyles(aux[1], map).replace('{tier}', 'small');
                auxMedium = addStyles(aux[1], map).replace('{tier}', 'medium');
                auxLarge = addStyles(aux[1], map).replace('{tier}', 'large');
                break;
            case 3:
                auxSmall = addStyles(aux[1], map).replace('{tier}', 'small');
                auxMedium = addStyles(aux[2], map).replace('{tier}', 'medium');
                auxLarge = addStyles(aux[2], map).replace('{tier}', 'large');
                break;
            default:
                auxSmall = addStyles(aux[1], map).replace('{tier}', 'small');
                auxMedium = addStyles(aux[2], map).replace('{tier}', 'medium');
                auxLarge = addStyles(aux[3], map).replace('{tier}', 'large');
                break;
        }
        output += ' ' + auxTiny.trim() + ' '
            + auxSmall.trim() + ' '
            + auxMedium.trim() + ' '
            + auxLarge.trim();
    }
    return output;
}
// Simple split
/**
 * @param {?} aux
 * @param {?=} char
 * @return {?}
 */
export function processSplit(aux, char) {
    if (char === void 0) { char = ','; }
    return aux.split(char).map((/**
     * @param {?} item
     * @return {?}
     */
    function (item) { return item.trim(); }));
}
/**
 * @param {?} el
 * @param {?} aux
 * @return {?}
 */
export function validateClassInElement(el, aux) {
    return el.hasAttribute(aux);
}
