// @allianz/ngx-ndbx/esm5/datefield/datepicker/calendar-body.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, NgZone, } from '@angular/core';
import { take } from 'rxjs/operators';
/**
 * An internal class that represents the data corresponding to a single calendar cell.
 * \@docs-private
 */
var /**
 * An internal class that represents the data corresponding to a single calendar cell.
 * \@docs-private
 */
NxCalendarCell = /** @class */ (function () {
    function NxCalendarCell(value, displayValue, ariaLabel, enabled) {
        this.value = value;
        this.displayValue = displayValue;
        this.ariaLabel = ariaLabel;
        this.enabled = enabled;
    }
    return NxCalendarCell;
}());
/**
 * An internal class that represents the data corresponding to a single calendar cell.
 * \@docs-private
 */
export { NxCalendarCell };
if (false) {
    /** @type {?} */
    NxCalendarCell.prototype.value;
    /** @type {?} */
    NxCalendarCell.prototype.displayValue;
    /** @type {?} */
    NxCalendarCell.prototype.ariaLabel;
    /** @type {?} */
    NxCalendarCell.prototype.enabled;
}
/**
 * An internal component used to display calendar data in a table.
 * \@docs-private
 */
var NxCalendarBodyComponent = /** @class */ (function () {
    function NxCalendarBodyComponent(_elementRef, _ngZone) {
        this._elementRef = _elementRef;
        this._ngZone = _ngZone;
        /**
         * The number of columns in the table.
         */
        this.numCols = 7;
        /**
         * Whether to allow selection of disabled cells.
         */
        this.allowDisabledSelection = false;
        /**
         * The cell number of the active cell in the table.
         */
        this.activeCell = 0;
        /**
         * The items to display in the first row in the offset space.
         */
        this.previousItems = [];
        /**
         * The items to display in the last row in the offset space.
         */
        this.followingItems = [[]];
        /**
         * Emits when a new value is selected.
         */
        this.selectedValueChange = new EventEmitter();
    }
    /**
     * @param {?} cell
     * @return {?}
     */
    NxCalendarBodyComponent.prototype._cellClicked = /**
     * @param {?} cell
     * @return {?}
     */
    function (cell) {
        if (!this.allowDisabledSelection && !cell.enabled) {
            return;
        }
        this.selectedValueChange.emit(cell.value);
    };
    Object.defineProperty(NxCalendarBodyComponent.prototype, "_firstRowOffset", {
        /** The number of blank cells to put at the beginning for the first row. */
        get: /**
         * The number of blank cells to put at the beginning for the first row.
         * @return {?}
         */
        function () {
            return this.rows && this.rows.length && this.rows[0].length ?
                this.numCols - this.rows[0].length : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxCalendarBodyComponent.prototype, "_lastRowOffset", {
        /** The number of blank cells to put at the end of the last filled row. */
        get: /**
         * The number of blank cells to put at the end of the last filled row.
         * @return {?}
         */
        function () {
            return this.rows && this.rows.length && this.rows[this.rows.length - 1].length ?
                this.numCols - this.rows[this.rows.length - 1].length : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxCalendarBodyComponent.prototype, "_lastRowIndex", {
        /** The index of the last row. */
        get: /**
         * The index of the last row.
         * @return {?}
         */
        function () {
            return this.rows && this.rows.length ?
                this.rows.length - 1 : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxCalendarBodyComponent.prototype, "_followingRows", {
        /**
         * The following full rows to display and fill up the calendar.
         */
        get: /**
         * The following full rows to display and fill up the calendar.
         * @return {?}
         */
        function () {
            if (this.followingItems && this.followingItems.length) {
                // if first row is not a full row => display them in offset space
                if (this.followingItems[0].length < this.numCols) {
                    return this.followingItems.slice(1, this.followingItems.length);
                }
                return this.followingItems;
            }
            return [];
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} rowIndex
     * @param {?} colIndex
     * @return {?}
     */
    NxCalendarBodyComponent.prototype._isActiveCell = /**
     * @param {?} rowIndex
     * @param {?} colIndex
     * @return {?}
     */
    function (rowIndex, colIndex) {
        /** @type {?} */
        var cellNumber = rowIndex * this.numCols + colIndex;
        // Account for the fact that the first row may not have as many cells.
        if (rowIndex) {
            cellNumber -= this._firstRowOffset;
        }
        return cellNumber === this.activeCell;
    };
    /** Focuses the active cell after the microtask queue is empty. */
    /**
     * Focuses the active cell after the microtask queue is empty.
     * @return {?}
     */
    NxCalendarBodyComponent.prototype._focusActiveCell = /**
     * Focuses the active cell after the microtask queue is empty.
     * @return {?}
     */
    function () {
        var _this = this;
        this._ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            _this._ngZone.onStable.asObservable().pipe(take(1)).subscribe((/**
             * @return {?}
             */
            function () {
                _this._elementRef.nativeElement.querySelector('.nx-calendar-body-active').focus();
            }));
        }));
    };
    NxCalendarBodyComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: '[nx-calendar-body]',
                    template: "<!-- Create the first and last row separately so we can include special spacer cells. -->\n<tr *ngFor=\"let row of rows; let rowIndex = index\" role=\"row\">\n  <!--\n    We mark this cell as aria-hidden so it doesn't get read out as one of the days in the week.\n    The aspect ratio of the table cells is maintained by setting the top and bottom padding as a\n    percentage of the width (a variant of the trick described here:\n    https://www.w3schools.com/howto/howto_css_aspect_ratio.asp).\n  -->\n  <ng-container *ngIf=\"rowIndex === 0\">\n    <td *ngFor=\"let item of previousItems\"\n        aria-hidden=\"true\"\n        class=\"nx-calendar-adjacent-cell\">\n      <div aria-hidden=\"true\"\n           class=\"nx-calendar-adjacent-cell-content\">\n        {{ item.displayValue }}\n      </div>\n    </td>\n  </ng-container>\n\n  <td *ngFor=\"let item of row; let colIndex = index\"\n      role=\"gridcell\"\n      class=\"nx-calendar-body-cell\"\n      [tabindex]=\"_isActiveCell(rowIndex, colIndex) ? 0 : -1\"\n      [class.nx-calendar-body-disabled]=\"!item.enabled\"\n      [class.nx-calendar-body-active]=\"_isActiveCell(rowIndex, colIndex)\"\n      [attr.aria-label]=\"item.ariaLabel\"\n      [attr.aria-disabled]=\"!item.enabled || null\"\n      (click)=\"_cellClicked(item)\"\n      [style.width.%]=\"100 / numCols\">\n    <div class=\"nx-calendar-body-cell-content\"\n         [class.nx-calendar-body-selected]=\"selectedValue === item.value\"\n         [class.nx-calendar-body-today]=\"todayValue === item.value\">\n      {{item.displayValue}}\n    </div>\n  </td>\n\n  <ng-container *ngIf=\"rowIndex === _lastRowIndex && _lastRowOffset\">\n    <td *ngFor=\"let item of followingItems[0]\"\n        aria-hidden=\"true\"\n        class=\"nx-calendar-adjacent-cell\">\n      <div aria-hidden=\"true\"\n           class=\"nx-calendar-adjacent-cell-content\">\n        {{ item.displayValue }}\n      </div>\n    </td>\n  </ng-container>\n\n</tr>\n\n<!-- \n  if there are still elements remaining in 'following-elements', then create some extra rows\n  to fill up the space\n-->\n<tr *ngFor=\"let row of _followingRows\"\n    aria-hidden=\"true\">\n  <td *ngFor=\"let item of row\"\n      class=\"nx-calendar-adjacent-cell\">\n    <div class=\"nx-calendar-adjacent-cell-content\">\n      {{ item.displayValue }}\n    </div>\n  </td>\n</tr>\n",
                    host: {
                        'class': 'nx-calendar-body',
                        'role': 'grid',
                        'attr.aria-readonly': 'true'
                    },
                    exportAs: 'nxCalendarBody',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".nx-calendar-adjacent-cell,.nx-calendar-body-cell{position:relative;height:0;line-height:0;text-align:center;outline:0;vertical-align:text-bottom}.nx-calendar-body-cell{cursor:pointer}.nx-calendar-body-disabled{cursor:default}.nx-calendar-body-cell-content{display:flex;align-items:center;justify-content:center;box-sizing:border-box;width:40px;height:40px;margin:0 auto;font-size:16px;font-weight:600;color:#414141;border-radius:50%;line-height:1;box-shadow:0 0 0 6px transparent}.nx-calendar-body-cell-content.nx-calendar-body-selected{background-color:#007ab3;color:#fff}.nx-calendar-body-disabled>.nx-calendar-body-cell-content:not(.nx-calendar-body-selected){color:#c2c2c2;font-weight:400}.nx-calendar-body-active:not(.nx-calendar-body-disabled)>.nx-calendar-body-cell-content:not(.nx-calendar-body-selected),:not(.nx-calendar-body-disabled):hover>.nx-calendar-body-cell-content:not(.nx-calendar-body-selected){background-color:#ececec;color:#414141}:host-context([data-whatinput=keyboard]) .nx-calendar-body-cell:focus .nx-calendar-body-cell-content{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6}:host-context([data-whatinput=keyboard]) .nx-calendar-body-cell:focus.nx-calendar-body-disabled .nx-calendar-body-cell-content{color:#c2c2c2}:host-context([data-whatinput=keyboard]) .nx-calendar-body-cell:focus .nx-calendar-body-cell-content:not(.nx-calendar-body-selected){background-color:#fff}.nx-calendar-adjacent-cell-content{display:flex;align-items:center;justify-content:center;box-sizing:border-box;width:40px;height:40px;margin:0 auto;font-size:16px;font-weight:400;color:#c2c2c2;border-radius:50%;line-height:1}"]
                }] }
    ];
    /** @nocollapse */
    NxCalendarBodyComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NgZone }
    ]; };
    NxCalendarBodyComponent.propDecorators = {
        label: [{ type: Input }],
        rows: [{ type: Input }],
        todayValue: [{ type: Input }],
        selectedValue: [{ type: Input }],
        numCols: [{ type: Input }],
        allowDisabledSelection: [{ type: Input }],
        activeCell: [{ type: Input }],
        previousItems: [{ type: Input }],
        followingItems: [{ type: Input }],
        selectedValueChange: [{ type: Output }]
    };
    return NxCalendarBodyComponent;
}());
export { NxCalendarBodyComponent };
if (false) {
    /**
     * The label for the table. (e.g. "Jan 2017").
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.label;
    /**
     * The cells to display in the table.
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.rows;
    /**
     * The value in the table that corresponds to today.
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.todayValue;
    /**
     * The value in the table that is currently selected.
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.selectedValue;
    /**
     * The number of columns in the table.
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.numCols;
    /**
     * Whether to allow selection of disabled cells.
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.allowDisabledSelection;
    /**
     * The cell number of the active cell in the table.
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.activeCell;
    /**
     * The items to display in the first row in the offset space.
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.previousItems;
    /**
     * The items to display in the last row in the offset space.
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.followingItems;
    /**
     * Emits when a new value is selected.
     * @type {?}
     */
    NxCalendarBodyComponent.prototype.selectedValueChange;
    /**
     * @type {?}
     * @private
     */
    NxCalendarBodyComponent.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    NxCalendarBodyComponent.prototype._ngZone;
}
