// @allianz/ngx-ndbx/esm5/table/sort-header/sort-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Optional, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { NxSortDirective } from './sort.directive';
import { ENTER, SPACE } from '@angular/cdk/keycodes';
import { NxSortHeaderIntl } from './sort-header-intl';
var NxSortHeaderComponent = /** @class */ (function () {
    function NxSortHeaderComponent(_sort, _intl, _changeDetectorRef) {
        var _this = this;
        this._sort = _sort;
        this._intl = _intl;
        this._changeDetectorRef = _changeDetectorRef;
        this._intlSubscription = this._intl.changes.subscribe((/**
         * @return {?}
         */
        function () { return _this._changeDetectorRef.markForCheck(); }));
    }
    Object.defineProperty(NxSortHeaderComponent.prototype, "key", {
        get: /**
         * @return {?}
         */
        function () {
            return this._key;
        },
        /** Sets the key of this sort header. */
        set: /**
         * Sets the key of this sort header.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this._key !== value) {
                this._key = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NxSortHeaderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._parentChangeSubscription = this._sort._stateChanges.subscribe((/**
         * @return {?}
         */
        function () {
            _this._changeDetectorRef.markForCheck();
        }));
    };
    /**
     * @return {?}
     */
    NxSortHeaderComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._parentChangeSubscription) {
            this._parentChangeSubscription.unsubscribe();
        }
        this._intlSubscription.unsubscribe();
    };
    /**
     * @return {?}
     */
    NxSortHeaderComponent.prototype._handleClick = /**
     * @return {?}
     */
    function () {
        this._sort.sort(this._key);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    NxSortHeaderComponent.prototype._onKeydown = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if ($event && ($event.keyCode === ENTER || $event.keyCode === SPACE)) {
            this._sort.sort(this._key);
            // prevent page from scrolling down
            if ($event.keyCode === SPACE) {
                $event.preventDefault();
            }
        }
    };
    /**
     * @return {?}
     */
    NxSortHeaderComponent.prototype._isSortedAscending = /**
     * @return {?}
     */
    function () {
        return (this._sort.active === this._key && this._sort.direction === 'asc');
    };
    /**
     * @return {?}
     */
    NxSortHeaderComponent.prototype._isSortedDescending = /**
     * @return {?}
     */
    function () {
        return (this._sort.active === this._key && this._sort.direction === 'desc');
    };
    /**
     * @return {?}
     */
    NxSortHeaderComponent.prototype._getAriaLabel = /**
     * @return {?}
     */
    function () {
        if (this._sort.active === this._key && this._sort.direction === 'asc') {
            return this._key + " " + this._intl.sortDescendingAriaLabel;
        }
        return this._key + " " + this._intl.sortAscendingAriaLabel;
    };
    NxSortHeaderComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'th[nxSortHeaderCell]',
                    exportAs: 'nxSortHeaderComponent',
                    template: "<div class=\"nx-sort-header__focus-container\">\n  <div class=\"nx-sort-header__icons-container\">\n    <nx-icon name=\"chevron-up-small\" class=\"nx-sort-header__icon-up\"\n            [class.nx-sort-header__hidden-icon]=\"_isSortedDescending()\">\n    </nx-icon>\n    <nx-icon name=\"chevron-down-small\" class=\"nx-sort-header__icon-down\"\n            [class.nx-sort-header__hidden-icon]=\"_isSortedAscending()\">\n    </nx-icon>\n  </div>\n  <ng-content></ng-content>\n</div>\n",
                    host: {
                        '(click)': '_handleClick()',
                        '(keydown)': '_onKeydown($event)',
                        '[tabindex]': '0',
                        'role': 'button',
                        '[attr.aria-label]': '_getAriaLabel()'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host{cursor:pointer}.nx-sort-header__icons-container{display:inline-block;margin-right:8px;vertical-align:bottom;width:24px;position:relative;height:24px}.nx-sort-header__icons-container nx-icon{font-size:20px;position:absolute;font-weight:600;width:24px}.nx-sort-header__icons-container .nx-sort-header__icon-up{top:-3px}.nx-sort-header__icons-container .nx-sort-header__icon-down{top:5px}.nx-sort-header__icons-container .nx-sort-header__hidden-icon{visibility:hidden}:host-context([data-whatinput=keyboard]):focus .nx-sort-header__focus-container{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;border-radius:4px}:host:focus{outline:0}"]
                }] }
    ];
    /** @nocollapse */
    NxSortHeaderComponent.ctorParameters = function () { return [
        { type: NxSortDirective, decorators: [{ type: Optional }] },
        { type: NxSortHeaderIntl },
        { type: ChangeDetectorRef }
    ]; };
    NxSortHeaderComponent.propDecorators = {
        key: [{ type: Input, args: ['nxSortHeaderCell',] }]
    };
    return NxSortHeaderComponent;
}());
export { NxSortHeaderComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxSortHeaderComponent.prototype._parentChangeSubscription;
    /**
     * @type {?}
     * @private
     */
    NxSortHeaderComponent.prototype._intlSubscription;
    /**
     * @type {?}
     * @private
     */
    NxSortHeaderComponent.prototype._key;
    /** @type {?} */
    NxSortHeaderComponent.prototype._sort;
    /** @type {?} */
    NxSortHeaderComponent.prototype._intl;
    /**
     * @type {?}
     * @private
     */
    NxSortHeaderComponent.prototype._changeDetectorRef;
}
