// @allianz/ngx-ndbx/esm5/radio-circle/radio-circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, EventEmitter, forwardRef, Input, Output, QueryList, Renderer2 } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { NxRadioCircleButton } from './radio-circle-button';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
/** @type {?} */
var nextId = 0;
/**
 * @deprecated Use `nx-circle-toggle-group` instead
 * \@deletion-target 8.0.0
 */
var NxRadioCircleComponent = /** @class */ (function () {
    function NxRadioCircleComponent(renderer) {
        this.renderer = renderer;
        this._id = nextId++;
        this._responsive = true;
        /**
         * An event emitted when the selection changes.
         */
        this.valueChange = new EventEmitter();
        /**
         * Whether the radio circle is disabled.
         */
        this.disabled = false;
        this.onTouchedCallback = (/**
         * @param {?} option
         * @return {?}
         */
        function (option) {
        });
        this.onChangeCallback = (/**
         * @param {?} option
         * @return {?}
         */
        function (option) {
        });
        console.warn('`nx-radio-circle` is deprecated. Use `nx-circle-toggle-group` instead.');
    }
    Object.defineProperty(NxRadioCircleComponent.prototype, "responsiveToggle", {
        /**
         * Whether the toggle behavior is disabled.
         *
         * @deprecated Use `nxResponsive` instead
         * @deletion-target 7.0.0
         */
        set: /**
         * Whether the toggle behavior is disabled.
         *
         * @deprecated Use `nxResponsive` instead
         * \@deletion-target 7.0.0
         * @param {?} value
         * @return {?}
         */
        function (value) {
            console.warn('`nxResponsiveToggle` is deprecated. Use `nxResponsive` instead.');
            this.responsive = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxRadioCircleComponent.prototype, "responsive", {
        get: /**
         * @return {?}
         */
        function () {
            return this._responsive;
        },
        /**
         * Whether the radio circle should have a responsive behavior.
         *
         * On false the circles will stay even on mobile viewports.
         * Default: true.
         */
        set: /**
         * Whether the radio circle should have a responsive behavior.
         *
         * On false the circles will stay even on mobile viewports.
         * Default: true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            /** @type {?} */
            var newValue = coerceBooleanProperty(value);
            if (newValue !== this.responsive) {
                this._responsive = newValue;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxRadioCircleComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selectedButton ? this._selectedButton.value : null;
        },
        /** Sets the value of the radio circle. */
        set: /**
         * Sets the value of the radio circle.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.writeValue(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @docs-private
     * @deprecated Use `responsive` getter
     * @deletion-target 7.0.0
     */
    /**
     * \@docs-private
     * @deprecated Use `responsive` getter
     * \@deletion-target 7.0.0
     * @return {?}
     */
    NxRadioCircleComponent.prototype.getResposiveToggle = /**
     * \@docs-private
     * @deprecated Use `responsive` getter
     * \@deletion-target 7.0.0
     * @return {?}
     */
    function () {
        console.warn('The method `getResponsiveToggle is deprecated. Use the `responsive` getter instead');
        return this._responsive;
    };
    Object.defineProperty(NxRadioCircleComponent.prototype, "radioCircleId", {
        /** The id of the radio circle. */
        get: /**
         * The id of the radio circle.
         * @return {?}
         */
        function () {
            return "radio-circle-" + this._id;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NxRadioCircleComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.renderer.addClass(this._buttons.first.labelElement.nativeElement, "nx-radio-circle__label--first");
        this.renderer.addClass(this._buttons.last.labelElement.nativeElement, "nx-radio-circle__label--last");
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NxRadioCircleComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this._buttons) {
            this._selectedButton = this._buttons.find((/**
             * @param {?} button
             * @return {?}
             */
            function (button) { return button.value === value; }));
        }
    };
    /**
     * @param {?} callback
     * @return {?}
     */
    NxRadioCircleComponent.prototype.registerOnChange = /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        this.onChangeCallback = callback;
    };
    /**
     * @param {?} callback
     * @return {?}
     */
    NxRadioCircleComponent.prototype.registerOnTouched = /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        this.onTouchedCallback = callback;
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    NxRadioCircleComponent.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
    };
    Object.defineProperty(NxRadioCircleComponent.prototype, "selectedButton", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            return this._selectedButton;
        },
        set: /**
         * @param {?} button
         * @return {?}
         */
        function (button) {
            if (this.disabled) {
                return;
            }
            this._selectedButton = button;
            this.onChangeCallback(this.value);
            this.valueChange.emit(this.value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxRadioCircleComponent.prototype, "buttons", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            return this._buttons;
        },
        enumerable: true,
        configurable: true
    });
    NxRadioCircleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-radio-circle',
                    template: "<div [id]=\"radioCircleId\" class=\"nx-radio-circle\" role=\"radiogroup\" [attr.aria-label]=\"name\">\n  <ng-content></ng-content>\n</div>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return NxRadioCircleComponent; })),
                            multi: true
                        }
                    ],
                    host: {
                        '[class.is-responsive]': 'responsive',
                        '[class.is-disabled]': 'disabled',
                        '[attr.aria-disabled]': 'disabled',
                    },
                    styles: [":host{display:block;position:relative}@media (max-width:703px){:host-context(.is-responsive) .nx-radio-circle{flex-direction:column}:host-context(.is-responsive) .nx-radio-circle__label{flex-direction:row;border:2px solid #007ab3;border-top-width:0;font-size:16px;padding:8px 40px 8px 15px}:host-context(.is-responsive) .nx-radio-circle__label--first{border-top-width:2px;border-radius:5px 5px 0 0}:host-context(.is-responsive) .nx-radio-circle__label--last{border-radius:0 0 5px 5px}:host-context(.is-responsive) .nx-radio-circle__circle{display:none}:host-context(.is-responsive) .nx-radio-circle__label-info{color:#007ab3;margin-left:10px;margin-top:0;line-height:16px;font-size:14px}:host-context(.is-responsive) .nx-radio-circle__mobile-check{display:inline-block;opacity:0;line-height:1;margin-right:10px}:host-context(.is-responsive) .nx-radio-circle--checked .nx-radio-circle__mobile-check{opacity:1}:host-context(.is-responsive) .nx-radio-circle--checked{background-color:#007ab3;color:#fff}:host-context(.is-responsive) .nx-radio-circle--checked .nx-radio-circle__label-info,:host-context(.is-responsive) .nx-radio-circle--checked .nx-radio-circle__label-text{color:#fff}:host(:not(.is-responsive)) ::ng-deep nx-radio-circle-icon-button,:host(:not(.is-responsive)) ::ng-deep nx-radio-circle-svg-button{flex:1 0 auto;min-width:33.3%}}.nx-radio-circle{display:flex;flex-direction:row;justify-content:space-around;flex-wrap:wrap}.nx-radio-circle__input{opacity:0;position:absolute}.nx-radio-circle__label{cursor:pointer;display:flex;flex-direction:column;justify-content:center;text-align:center;align-items:center;font-size:16px;font-weight:600;padding:0;position:relative;transition:background-color .2s,border-color .2s,color .2s;vertical-align:middle}@media (min-width:704px){:host:hover .nx-radio-circle__circle{background-color:#006192;border-color:transparent;box-shadow:none;color:#fff}.nx-radio-circle__label{line-height:8rem;min-height:8rem;min-width:8rem;max-width:8rem;padding:0}}.nx-radio-circle__label-text{line-height:20px;z-index:1;color:#414141}@media (max-width:703px){.nx-radio-circle__label-text{color:#007ab3}}.nx-radio-circle__input:focus+.nx-radio-circle__label{color:#fff}.nx-radio-circle__input:focus+.nx-radio-circle__label:after{background-color:#006192;border-color:transparent}.nx-radio-circle__mobile-check{font-size:24px;display:none;z-index:1}.nx-radio-circle--checked .nx-radio-circle__circle{color:#fff;background-color:#007ab3}.nx-radio-circle--checked{color:#fff}.nx-radio-circle--checked:hover{cursor:default}@media screen and (-ms-high-contrast:active){.nx-radio-circle__label-text{color:#fff}.nx-radio-circle--checked .nx-radio-circle__circle{background-color:#ff0;border-color:#ff0}.nx-radio-circle--checked,.nx-radio-circle--checked .nx-radio-circle__label-text{color:#000}.nx-radio-circle--checked .nx-radio-circle__label-text{color:#fff}}.nx-radio-circle__circle{display:flex;justify-content:center;align-items:center;border-radius:50%;min-width:8rem;min-height:8rem;max-width:8rem;max-height:8rem;width:8rem;height:8rem;border:2px solid #007ab3;margin-bottom:1.5rem;color:#007ab3}:host.is-disabled *{color:#c2c2c2}:host.is-disabled .nx-radio-circle__circle{border:2px solid #c2c2c2}:host.is-disabled .nx-radio-circle--checked .nx-radio-circle__circle{background-color:#c2c2c2;border:2px solid #c2c2c2}:host.is-disabled .nx-radio-circle--checked .nx-radio-circle__label--icon{color:#fff!important}@media (max-width:703px){:host.is-disabled .nx-radio-circle__label{border:2px solid #c2c2c2}:host.is-disabled .nx-radio-circle__label-info{color:#c2c2c2}:host.is-disabled .nx-icon--check:before{color:#fff}:host.is-disabled .nx-radio-circle--checked{background-color:#c2c2c2}:host.is-disabled .nx-radio-circle--checked .nx-radio-circle__label-info{color:#fff}}:host.is-disabled:hover{cursor:not-allowed}:host.is-disabled:hover .nx-radio-circle__circle{background-color:transparent;cursor:not-allowed}.nx-radio-circle__label--icon{pointer-events:none;z-index:1;display:flex;text-align:center;align-items:center;justify-content:center;font-size:4rem;font-weight:400;position:relative}.nx-radio-circle__label--icon.nx-icon--auto{font-size:4rem}.nx-radio-circle__label-info{color:#fff;display:inline-block;font-weight:300;line-height:22px;font-size:18px;margin-top:4px;transition:color .2s;z-index:1}@media screen and (-ms-high-contrast:active){.nx-radio-circle__label-info{color:#fff}}.nx-radio-circle__label-info:before{content:'(';display:inline-block}@media (min-width:704px){.nx-radio-circle__label-info{color:#767676;display:block;width:16rem}.nx-radio-circle__label-info:before{content:''}}.nx-radio-circle__label-info:after{content:')';display:inline-block}@media (min-width:704px){.nx-radio-circle__label-info:after{content:''}}:host-context([data-whatinput=keyboard]) .nx-radio-circle__input:focus+.nx-radio-circle__label .nx-radio-circle__circle{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6}"]
                }] }
    ];
    /** @nocollapse */
    NxRadioCircleComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    NxRadioCircleComponent.propDecorators = {
        _buttons: [{ type: ContentChildren, args: [NxRadioCircleButton,] }],
        valueChange: [{ type: Output, args: ['nxValueChange',] }],
        name: [{ type: Input, args: ['nxName',] }],
        disabled: [{ type: Input, args: ['disabled',] }],
        responsiveToggle: [{ type: Input, args: ['nxResponsiveToggle',] }],
        responsive: [{ type: Input, args: ['nxResponsive',] }],
        value: [{ type: Input, args: ['nxValue',] }]
    };
    return NxRadioCircleComponent;
}());
export { NxRadioCircleComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxRadioCircleComponent.prototype._id;
    /**
     * @type {?}
     * @private
     */
    NxRadioCircleComponent.prototype._buttons;
    /**
     * @type {?}
     * @private
     */
    NxRadioCircleComponent.prototype._selectedButton;
    /**
     * @type {?}
     * @private
     */
    NxRadioCircleComponent.prototype._responsive;
    /**
     * An event emitted when the selection changes.
     * @type {?}
     */
    NxRadioCircleComponent.prototype.valueChange;
    /**
     * Sets the name of the radio circle, which is used for accessibility.
     * @type {?}
     */
    NxRadioCircleComponent.prototype.name;
    /**
     * Whether the radio circle is disabled.
     * @type {?}
     */
    NxRadioCircleComponent.prototype.disabled;
    /**
     * @type {?}
     * @private
     */
    NxRadioCircleComponent.prototype.onTouchedCallback;
    /**
     * @type {?}
     * @private
     */
    NxRadioCircleComponent.prototype.onChangeCallback;
    /**
     * @type {?}
     * @private
     */
    NxRadioCircleComponent.prototype.renderer;
}
