// @allianz/ngx-ndbx/esm5/radio-circle/radio-circle-button.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef, Input, ViewChild } from '@angular/core';
/** @type {?} */
var nextId = 0;
/**
 * @abstract
 */
var NxRadioCircleButton = /** @class */ (function () {
    function NxRadioCircleButton(radioCircle) {
        this.radioCircle = radioCircle;
        /**
         * \@docs-private
         */
        this.id = ++nextId;
    }
    /**
     * @return {?}
     */
    NxRadioCircleButton.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        if (!this.label) {
            console.warn("nx-radio-circle-icon-button/nx-radio-circle-svg-button: required field 'nxLabel' is not set");
        }
    };
    Object.defineProperty(NxRadioCircleButton.prototype, "checked", {
        get: /**
         * @return {?}
         */
        function () {
            return this.radioCircle.selectedButton === this || null;
        },
        /** Whether the item is preselected. */
        set: /**
         * Whether the item is preselected.
         * @param {?} checked
         * @return {?}
         */
        function (checked) {
            if (checked) {
                this.radioCircle.selectedButton = this;
            }
            else {
                if (this.checked) {
                    this.radioCircle.selectedButton = null;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxRadioCircleButton.prototype, "circleButtonId", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            return this.radioCircle.radioCircleId + '-' + this.id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxRadioCircleButton.prototype, "radioCircleId", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            return this.radioCircle.radioCircleId;
        },
        enumerable: true,
        configurable: true
    });
    NxRadioCircleButton.propDecorators = {
        label: [{ type: Input, args: ['nxLabel',] }],
        hint: [{ type: Input, args: ['nxHint',] }],
        value: [{ type: Input, args: ['nxValue',] }],
        labelElement: [{ type: ViewChild, args: ['labelElement',] }],
        checked: [{ type: Input, args: ['nxSelected',] }]
    };
    return NxRadioCircleButton;
}());
export { NxRadioCircleButton };
if (false) {
    /**
     * \@docs-private
     * @type {?}
     */
    NxRadioCircleButton.prototype.id;
    /**
     * Sets the label displayed below the circle.
     * @type {?}
     */
    NxRadioCircleButton.prototype.label;
    /**
     * Sets an additional hint displayed below the label.
     * @type {?}
     */
    NxRadioCircleButton.prototype.hint;
    /**
     * Sets the value used in the model.
     * @type {?}
     */
    NxRadioCircleButton.prototype.value;
    /**
     * \@docs-private
     * @type {?}
     */
    NxRadioCircleButton.prototype.labelElement;
    /**
     * \@docs-private
     * @type {?}
     * @protected
     */
    NxRadioCircleButton.prototype.radioCircle;
}
