// @allianz/ngx-ndbx/esm5/base/error/error.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, InjectionToken, Inject, Optional } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
/**
 * Represents the default options for the error notification that can be configured
 * using the `ERROR_DEFAULT_OPTIONS` injection token.
 * @record
 */
export function ErrorDefaultOptions() { }
if (false) {
    /**
     * Defines the style type of the error notification.
     * @type {?|undefined}
     */
    ErrorDefaultOptions.prototype.appearance;
}
/** @type {?} */
export var ERROR_DEFAULT_OPTIONS = new InjectionToken('ERROR_DEFAULT_OPTIONS');
var NxErrorComponent = /** @class */ (function () {
    function NxErrorComponent(_changeDetectorRef, _defaultOptions) {
        this._changeDetectorRef = _changeDetectorRef;
        this._defaultOptions = _defaultOptions;
        this._showIcon = true;
        this._appearance = 'message';
        if (this._defaultOptions && this._defaultOptions.appearance) {
            this.appearance = this._defaultOptions.appearance;
        }
    }
    Object.defineProperty(NxErrorComponent.prototype, "showIcon", {
        get: /**
         * @return {?}
         */
        function () {
            return this._showIcon;
        },
        /** Whether an icon should be displayed. */
        set: /**
         * Whether an icon should be displayed.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._showIcon = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxErrorComponent.prototype, "appearance", {
        get: /**
         * @return {?}
         */
        function () {
            return this._appearance;
        },
        /**
         * Whether the error should have message or text styling.
         *
         * Default is 'message'.
         */
        set: /**
         * Whether the error should have message or text styling.
         *
         * Default is 'message'.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value !== this.appearance) {
                this._appearance = value;
                this._changeDetectorRef.markForCheck();
            }
        },
        enumerable: true,
        configurable: true
    });
    NxErrorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-error',
                    template: "<!-- b2b (expert) error notification -->\n<ng-container *ngIf=\"appearance === 'text'\">\n  <nx-icon *ngIf=\"showIcon\" class=\"nx-error__icon\" [name]=\"'exclamation-triangle'\">\n  </nx-icon>\n  <div class=\"nx-error__content-wrapper\">\n    <div class=\"nx-error__content\">\n      <ng-container [ngTemplateOutlet]=\"errorContent\"></ng-container>\n    </div>\n  </div>\n</ng-container>\n\n<!-- b2c error notification -->\n<nx-message *ngIf=\"appearance === 'message'\" [showIcon]=\"showIcon\" nxContext=\"error\">\n  <ng-container [ngTemplateOutlet]=\"errorContent\"></ng-container>\n</nx-message>\n\n<ng-template #errorContent>\n  <ng-content></ng-content>\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[attr.role]': '"alert"'
                    },
                    styles: [":host{font-size:16px;line-height:20px;color:#dc3149;width:100%;display:block}:host .nx-error__content{line-height:20px}:host .nx-error__icon{font-size:20px;line-height:20px;color:#dc3149;margin-right:8px;float:left}"]
                }] }
    ];
    /** @nocollapse */
    NxErrorComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [ERROR_DEFAULT_OPTIONS,] }] }
    ]; };
    NxErrorComponent.propDecorators = {
        showIcon: [{ type: Input }],
        appearance: [{ type: Input }]
    };
    return NxErrorComponent;
}());
export { NxErrorComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxErrorComponent.prototype._showIcon;
    /**
     * @type {?}
     * @private
     */
    NxErrorComponent.prototype._appearance;
    /**
     * @type {?}
     * @private
     */
    NxErrorComponent.prototype._changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    NxErrorComponent.prototype._defaultOptions;
}
