// @allianz/ngx-ndbx/esm2015/table/table.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
/**
 * This is a table.
 * The table supports a `zebra` and a `condensed` mode.
 */
export class NxTableComponent {
    /**
     * @param {?} _changeDetectorRef
     */
    constructor(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
    }
    /**
     * Change the table mode to condensed
     * @param {?} value
     * @return {?}
     */
    set condensed(value) {
        /** @type {?} */
        const newValue = coerceBooleanProperty(value);
        if (value !== this._condensed) {
            this._condensed = newValue;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get condensed() {
        return this._condensed;
    }
    /**
     * Change the table mode to zebra
     * @param {?} value
     * @return {?}
     */
    set zebra(value) {
        /** @type {?} */
        const newValue = coerceBooleanProperty(value);
        if (value !== this._zebra) {
            this._zebra = newValue;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get zebra() {
        return this._zebra;
    }
}
NxTableComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'table[nxTable]',
                template: '<ng-content></ng-content>',
                host: {
                    'class': 'nx-table',
                    '[class.nx-table--condensed]': 'condensed',
                    '[class.nx-table--zebra]': 'zebra'
                },
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{border-spacing:0;border-bottom:2px solid #414141;width:100%}::ng-deep .nx-header-cell{font-weight:600;line-height:24px;border-top:1px solid #d9d9d9;border-bottom:2px solid #414141;text-align:left;white-space:nowrap;padding:16px 16px 14px}::ng-deep .nx-header-cell:first-child{padding-left:8px}::ng-deep .nx-header-cell:last-child{padding-right:8px}::ng-deep .nx-table-cell{line-height:24px;border-top:1px solid #d9d9d9;padding:16px 16px 15px}::ng-deep .nx-table-cell:first-child{padding-left:8px}::ng-deep .nx-table-cell:last-child{padding-right:8px}:host(.nx-table--condensed) ::ng-deep .nx-table-cell{padding-bottom:7px;padding-top:8px}:host(.nx-table--zebra) ::ng-deep .nx-table-cell{border-top:1px solid #f5f5f5}:host(.nx-table--zebra) ::ng-deep .nx-table-cell--expandable{border-top:0}:host(.nx-table--condensed) ::ng-deep .nx-table-cell--expandable{padding-bottom:0;padding-top:0}:host(.nx-table--condensed) ::ng-deep .nx-table-cell--expandable .nx-table-cell__wrapper{padding-bottom:7px;padding-top:8px}::ng-deep .nx-expandable-table-cell__content{overflow:hidden}::ng-deep .nx-expandable-table-cell__wrapper{padding:16px}::ng-deep .nx-expandable-table-cell:first-child .nx-cell__wrapper{padding-left:8px}::ng-deep .nx-expandable-table-cell:last-child .nx-cell__wrapper{padding-right:8px}:host(.nx-table--condensed) ::ng-deep .nx-expandable-table-cell .nx-expandable-table-cell__wrapper{padding-bottom:7px;padding-top:8px}::ng-deep .nx-table-row:first-child .nx-table-cell{border-top:0}:host(.nx-table--zebra) ::ng-deep .nx-table-row:nth-child(2n){background-color:#f5f5f5}:host(.nx-table--zebra) ::ng-deep .nx-expandable-table-row:nth-child(4n){background-color:#f5f5f5}:host(.nx-table--zebra) ::ng-deep .nx-expandable-table-row:nth-child(4n)+.nx-table-row{background-color:#fff}:host(.nx-table--zebra) ::ng-deep .nx-expandable-table-row:nth-child(4n+2){background-color:#fff}:host(.nx-table--zebra) ::ng-deep .nx-expandable-table-row:nth-child(4n+2)+.nx-table-row{background-color:#f5f5f5}::ng-deep .nx-exandable-table-cell nx-checkbox,::ng-deep .nx-header-cell nx-checkbox,::ng-deep .nx-table-cell nx-checkbox{margin:0}::ng-deep .nx-exandable-table-cell nx-checkbox .nx-checkbox__label,::ng-deep .nx-header-cell nx-checkbox .nx-checkbox__label,::ng-deep .nx-table-cell nx-checkbox .nx-checkbox__label{padding:0}::ng-deep .nx-exandable-table-cell nx-taglist,::ng-deep .nx-header-cell nx-taglist,::ng-deep .nx-table-cell nx-taglist{margin:0}::ng-deep .nx-exandable-table-cell nx-taglist .nx-taglist__list-item,::ng-deep .nx-header-cell nx-taglist .nx-taglist__list-item,::ng-deep .nx-table-cell nx-taglist .nx-taglist__list-item{margin-bottom:0}::ng-deep .nx-exandable-table-cell button[nxbutton],::ng-deep .nx-header-cell button[nxbutton],::ng-deep .nx-table-cell button[nxbutton]{margin:0}::ng-deep .nx-exandable-table-cell nx-formfield .nx-formfield__wrapper,::ng-deep .nx-header-cell nx-formfield .nx-formfield__wrapper,::ng-deep .nx-table-cell nx-formfield .nx-formfield__wrapper{padding:0}::ng-deep .nx-exandable-table-cell [nxlist],::ng-deep .nx-header-cell [nxlist],::ng-deep .nx-table-cell [nxlist]{margin:0}::ng-deep .nx-exandable-table-cell [nxlist] li:last-child,::ng-deep .nx-header-cell [nxlist] li:last-child,::ng-deep .nx-table-cell [nxlist] li:last-child{margin-bottom:0}"]
            }] }
];
/** @nocollapse */
NxTableComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
NxTableComponent.propDecorators = {
    condensed: [{ type: Input }],
    zebra: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxTableComponent.prototype._condensed;
    /**
     * @type {?}
     * @private
     */
    NxTableComponent.prototype._zebra;
    /**
     * @type {?}
     * @private
     */
    NxTableComponent.prototype._changeDetectorRef;
}
