// @allianz/ngx-ndbx/esm5/tabs/tab-body.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { CdkPortalOutlet, TemplatePortal } from '@angular/cdk/portal';
import { ChangeDetectionStrategy, Component, Input, ViewChild } from '@angular/core';
/**
 * \@docs-private
 */
var NxTabBodyComponent = /** @class */ (function () {
    function NxTabBodyComponent() {
        this._active = false;
    }
    Object.defineProperty(NxTabBodyComponent.prototype, "active", {
        get: /**
         * @return {?}
         */
        function () {
            return this._active;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._active = coerceBooleanProperty(value);
            if (this._active) {
                this.attach();
            }
            else {
                this.detach();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NxTabBodyComponent.prototype.attach = /**
     * @return {?}
     */
    function () {
        if (!this.portalOutlet.hasAttached()) {
            this.portalOutlet.attach(this.content);
        }
    };
    /**
     * @return {?}
     */
    NxTabBodyComponent.prototype.detach = /**
     * @return {?}
     */
    function () {
        if (this.portalOutlet.hasAttached()) {
            this.portalOutlet.detach();
        }
    };
    NxTabBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-tab-body',
                    template: "<ng-template cdkPortalOutlet></ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host{display:block}"]
                }] }
    ];
    NxTabBodyComponent.propDecorators = {
        content: [{ type: Input }],
        active: [{ type: Input }],
        portalOutlet: [{ type: ViewChild, args: [CdkPortalOutlet,] }]
    };
    return NxTabBodyComponent;
}());
export { NxTabBodyComponent };
if (false) {
    /** @type {?} */
    NxTabBodyComponent.prototype.content;
    /** @type {?} */
    NxTabBodyComponent.prototype.portal;
    /**
     * @type {?}
     * @private
     */
    NxTabBodyComponent.prototype._active;
    /** @type {?} */
    NxTabBodyComponent.prototype.portalOutlet;
}
