// @allianz/ngx-ndbx/esm2015/progressbar/progressbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, ChangeDetectionStrategy } from '@angular/core';
import { Component } from '@angular/core';
import { clamp } from '@allianz/ngx-ndbx/utils';
/** @type {?} */
let progressbarId = 0;
export class NxProgressbarComponent {
    constructor() {
        /**
         * \@docs-private
         */
        this.progressbarId = `nx-progress-bar-${progressbarId++}`;
        this._value = 0;
    }
    /**
     * Sets the value of the progress bar. Defaults to zero. Mirrored to aria-value now.
     * @return {?}
     */
    get value() { return this._value; }
    /**
     * @param {?} value
     * @return {?}
     */
    set value(value) { this._value = clamp(value || 0); }
    /**
     * @return {?}
     */
    _primaryTransform() {
        /** @type {?} */
        const scale = this.value;
        return { transform: `scaleX(${scale})` };
    }
}
NxProgressbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-progressbar',
                template: "<div class=\"nx-progress__track\">\n  <div class=\"nx-progress__indicator\" [ngStyle]=\"_primaryTransform()\"></div>\n</div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[attr.aria-valuenow]': 'value'
                },
                styles: [":host{display:block;height:4px;overflow:hidden;position:relative;transition:opacity 250ms linear;width:100%}.nx-progress__track{background-color:#ececec;height:100%}.nx-progress__indicator{-webkit-transform-origin:top left;transform-origin:top left;height:100%;position:absolute;width:100%}.nx-progress__indicator::after{height:100%;position:absolute;width:100%;-webkit-animation:none;animation:none;content:'';display:inline-block;left:0;background:#007ab3}"]
            }] }
];
NxProgressbarComponent.propDecorators = {
    value: [{ type: Input }]
};
if (false) {
    /**
     * \@docs-private
     * @type {?}
     */
    NxProgressbarComponent.prototype.progressbarId;
    /**
     * @type {?}
     * @private
     */
    NxProgressbarComponent.prototype._value;
}
