// @allianz/ngx-ndbx/esm2015/tree/tree.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CdkTree } from '@angular/cdk/tree';
import { ChangeDetectionStrategy, Component, ViewChild, ViewEncapsulation } from '@angular/core';
import { NxTreeNodeOutletDirective } from './outlet';
/**
 * Wrapper for the CdkTable with NDBX design styles.
 * @template T
 */
export class NxTreeComponent extends CdkTree {
}
NxTreeComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-tree',
                exportAs: 'nxTree',
                template: `<ng-container nxTreeNodeOutlet></ng-container>`,
                host: {
                    'class': 'nx-tree',
                    'role': 'tree',
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [{ provide: CdkTree, useExisting: NxTreeComponent }],
                styles: [":host.nx-tree{display:block}.nx-tree__node{display:block;overflow:hidden}"]
            }] }
];
NxTreeComponent.propDecorators = {
    _nodeOutlet: [{ type: ViewChild, args: [NxTreeNodeOutletDirective,] }]
};
if (false) {
    /** @type {?} */
    NxTreeComponent.prototype._nodeOutlet;
}
