// @allianz/ngx-ndbx/esm2015/progress-stepper/multi/multi-step-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { coerceBooleanProperty, coerceNumberProperty } from '@angular/cdk/coercion';
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef, forwardRef, Inject } from '@angular/core';
import { NxStepComponent, NxProgressStepperDirective } from './../progress-stepper.component';
import { ErrorStateMatcher } from '@allianz/ngx-ndbx/utils';
/**
 * \@docs-private
 */
export class NxMultiStepItemComponent extends NxStepComponent {
    /**
     * @param {?} _changeDetectorRef
     * @param {?} stepper
     * @param {?} _errorStateMatcher
     */
    constructor(_changeDetectorRef, stepper, _errorStateMatcher) {
        super(stepper, _errorStateMatcher);
        this._changeDetectorRef = _changeDetectorRef;
        this._completed = false;
    }
    /**
     * Sets the index of the selected step.
     * @return {?}
     */
    get index() { return this._index; }
    /**
     * @param {?} value
     * @return {?}
     */
    set index(value) { this._index = coerceNumberProperty(value); }
    /**
     * Sets the selected step.
     * @return {?}
     */
    get selected() { return this._selected; }
    /**
     * @param {?} value
     * @return {?}
     */
    set selected(value) { this._selected = coerceBooleanProperty(value); }
    /**
     * Sets the active step.
     * @return {?}
     */
    get active() { return this._active; }
    /**
     * @param {?} value
     * @return {?}
     */
    set active(value) { this._active = coerceBooleanProperty(value); }
    /**
     * Sets the last step.
     * @return {?}
     */
    get last() { return this._last; }
    /**
     * @param {?} value
     * @return {?}
     */
    set last(value) { this._last = coerceBooleanProperty(value); }
    /**
     * Sets the first step.
     * @return {?}
     */
    get first() { return this._first; }
    /**
     * @param {?} value
     * @return {?}
     */
    set first(value) { this._first = coerceBooleanProperty(value); }
    /**
     * Whether the step is completed or not.
     * @param {?} value
     * @return {?}
     */
    set stepCompleted(value) {
        this._completed = coerceBooleanProperty(value);
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get stepCompleted() { return this._completed; }
}
NxMultiStepItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-multi-step-item',
                template: "<div class=\"label\">\n    {{label}}\n  </div>\n\n<div\n  [ngSwitch]=\"state\"\n  class=\"dot\"\n  [class.dot--filled]='stepCompleted'>\n  <nx-icon *ngIf=\"stepCompleted\" name=\"check\"></nx-icon>\n  <div *ngIf=\"selected && !stepCompleted\" class=\"small-dot\"></div>\n\n</div>\n<div\n  *ngIf=\"!last\"\n  class=\"before-line\"\n  [class.before-line--active]='active'>\n</div>\n<div\n  *ngIf=\"!last\"\n  class=\"line\"\n  [class.line--active]='active'>\n</div>\n<div\n  *ngIf=\"!last\"\n  class=\"after-line\"\n  [class.after-line--active]='active'>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{flex:1 0 auto;display:flex;align-items:center;position:relative;font-size:16px;padding-bottom:32px;cursor:pointer}.label{position:absolute;top:calc(8px + 16px);-webkit-transform:translateX(-50%) translateX(8px);transform:translateX(-50%) translateX(8px);min-width:156px;text-align:center;line-height:24px;font-size:16px;color:#414141}.dot{width:16px;height:16px;border:2px solid #007ab3;border-radius:50%;color:#fff;background-color:#fff;flex:0 0 auto}.dot nx-icon{font-size:16px;position:absolute;top:0;left:0}.small-dot{top:4px;left:4px;position:absolute;width:8px;height:8px;border-radius:50%;background-color:#007ab3}.dot--filled{background-color:#007ab3;color:#fff}.line{flex:1 1 100%;min-width:32px;height:0;border-top:4px dotted #c2c2c2}.line--active{border-top-color:#007ab3;border-top-style:solid;border-image-source:none;transition:border-style .1s ease-in}.after-line,.before-line{width:4px;height:4px;z-index:-1}.after-line--active,.before-line--active{z-index:0;background:#007ab3}"]
            }] }
];
/** @nocollapse */
NxMultiStepItemComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: NxProgressStepperDirective, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => NxProgressStepperDirective)),] }] },
    { type: ErrorStateMatcher }
];
NxMultiStepItemComponent.propDecorators = {
    state: [{ type: Input }],
    label: [{ type: Input }],
    index: [{ type: Input }],
    selected: [{ type: Input }],
    active: [{ type: Input }],
    last: [{ type: Input }],
    first: [{ type: Input }],
    stepCompleted: [{ type: Input }]
};
if (false) {
    /**
     * The state of the step.
     * @type {?}
     */
    NxMultiStepItemComponent.prototype.state;
    /**
     * The label of the step.
     * @type {?}
     */
    NxMultiStepItemComponent.prototype.label;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._index;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._selected;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._active;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._last;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._first;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._completed;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._changeDetectorRef;
}
