// @allianz/ngx-ndbx/esm2015/checkbox-circle/checkbox-circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, forwardRef, Input, Output } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/** @type {?} */
let nextId = 0;
/**
 * @deprecated Use `nx-circle-toggle` instead
 * \@deletion-target 8.0.0
 */
export class NxCheckboxCircleComponent {
    constructor() {
        this._id = (nextId++).toString();
        /**
         * An event emitted when the checked value has changed.
         */
        this.checkedChange = new EventEmitter(false);
        this.onChangeCallback = (/**
         * @param {?} checked
         * @return {?}
         */
        (checked) => { });
        this.onTouchedCallback = (/**
         * @return {?}
         */
        () => { });
        console.warn('`nx-checkbox-circle` is deprecated. Use `nx-circle-toggle` instead.');
    }
    /**
     * Id of the checkbox.
     * If not set, the checkbox gets an incremented value by default.
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        if (value !== this._id) {
            this._id = value;
        }
    }
    /**
     * @return {?}
     */
    get id() {
        return `nx-checkbox-circle-${this._id}`;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.checked = value;
        this.checkedChange.emit(value);
    }
    /**
     * @param {?} onChange
     * @return {?}
     */
    registerOnChange(onChange) {
        this.onChangeCallback = onChange;
    }
    /**
     * @param {?} onTouched
     * @return {?}
     */
    registerOnTouched(onTouched) {
        this.onTouchedCallback = onTouched;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
    }
    /**
     * \@docs-private
     * @return {?}
     */
    toggle() {
        this.writeValue(!this.checked);
        this.onChangeCallback(this.checked);
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get svgButton() {
        return !!this.svg && !!this.svgChecked;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get svgUrl() {
        return this.checked ? this.svgChecked : this.svg;
    }
}
NxCheckboxCircleComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-checkbox-circle',
                template: "<input [id]=\"id\" class=\"nx-checkbox-circle__input\" [name]=\"name\" [attr.aria-labelledby]=\"id + '-label'\"\n    type=\"checkbox\" [checked]=\"checked\" (click)=\"toggle()\">\n<label class=\"nx-checkbox-circle__label\" [for]=\"id\"\n  [id]=\"id + '-label'\">\n  <span class=\"nx-checkbox-circle__check-icon\">\n    <nx-icon name=\"check-circle\"></nx-icon>\n  </span>\n  <span class=\"nx-checkbox-circle__icon\">\n    <nx-icon *ngIf=\"iconName\" [name]=\"iconName\"></nx-icon>\n    <img *ngIf=\"svgButton\" class=\"nx-checkbox-circle__icon-image\" [src]=\"svgUrl\"\n    />\n  </span>\n  <span class=\"nx-checkbox-circle__label-text\">\n    {{label}}\n  </span>\n</label>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => NxCheckboxCircleComponent)),
                        multi: true
                    }
                ],
                host: {
                    '[class.is-checked]': 'checked'
                },
                styles: [":host{display:inline-block;margin-right:65px;vertical-align:top;position:relative}.nx-checkbox-circle__icon-image{height:100%}.nx-checkbox-circle__label-text{display:inline-block;max-width:16rem}.nx-checkbox-circle__input{opacity:0;position:absolute}.nx-checkbox-circle__check-icon{background-color:#f5f5f5;border-radius:50%;color:#3da556;display:block;font-size:31px;text-align:center;line-height:1;width:32px;height:32px;margin-left:1.6rem;padding:0;position:absolute;top:0;left:50%;opacity:0;transition:opacity .3s;z-index:1}@media (max-width:703px){:host{margin-right:48px}.nx-checkbox-circle__check-icon{width:28px;height:28px;font-size:27px;left:45%}}:host(.is-checked) .nx-checkbox-circle__check-icon{opacity:1;transition:opacity .3s}.nx-checkbox-circle__icon{border:2px solid #007ab3;border-radius:50%;content:\"\";display:block;height:8rem;width:8rem;left:50%;margin-left:-4rem;position:absolute;top:5px;transition:background-color .2s,border-color .2s,box-shadow .2s}.nx-checkbox-circle__icon nx-icon{border-radius:50%;color:#007ab3;display:block;font-size:4.8rem;position:absolute;text-align:center;margin:1.4rem;z-index:1}.nx-checkbox-circle__icon:hover{border-color:#006192;cursor:pointer}.nx-checkbox-circle__icon:hover nx-icon{color:#006192}:host(.is-checked) .nx-checkbox-circle__icon{background-color:#007ab3}@media screen and (-ms-high-contrast:active){:host(.is-checked) .nx-checkbox-circle__icon{color:#000}}:host(.is-checked) .nx-checkbox-circle__icon:hover{background-color:#006192;border-color:#006192}:host(.is-checked) .nx-checkbox-circle__icon nx-icon{color:#fff}.nx-checkbox-circle__label{display:inline-block;font-size:16px;font-weight:600;line-height:20px;margin-right:8px;padding:9.5rem 0 0;position:relative;text-align:center}@media (max-width:703px){.nx-checkbox-circle__icon{width:6.4rem;height:6.4rem;margin-left:-3.2rem}.nx-checkbox-circle__icon nx-icon{margin:1.1rem;font-size:4rem}.nx-checkbox-circle__label{margin-right:6px;padding:8rem 0 0;font-size:14px;line-height:16px}}.nx-checkbox-circle__text{display:block;position:absolute;text-align:center;left:0;top:5px;font-size:22px;line-height:8rem;width:8rem;height:8rem;color:#007ab3;border-radius:50%;border:2px solid #007ab3;z-index:0;cursor:pointer}.nx-checkbox-circle__text:hover{color:#006192;border-color:#006192}@media (max-width:703px){.nx-checkbox-circle__text{font-size:18px;line-height:6.4rem;width:6.4rem;height:6.4rem}}:host-context([data-whatinput=keyboard]) .nx-checkbox-circle__input:focus+.nx-checkbox-circle__label ::ng-deep .nx-checkbox-circle__icon{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6}"]
            }] }
];
/** @nocollapse */
NxCheckboxCircleComponent.ctorParameters = () => [];
NxCheckboxCircleComponent.propDecorators = {
    id: [{ type: Input }],
    name: [{ type: Input }],
    checkedChange: [{ type: Output }],
    checked: [{ type: Input }],
    iconName: [{ type: Input, args: ['nxIcon',] }],
    svg: [{ type: Input, args: ['nxSvg',] }],
    svgChecked: [{ type: Input, args: ['nxSvgChecked',] }],
    label: [{ type: Input, args: ['nxLabel',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxCheckboxCircleComponent.prototype._id;
    /**
     * The name of the Checkbox Circle.
     * @type {?}
     */
    NxCheckboxCircleComponent.prototype.name;
    /**
     * An event emitted when the checked value has changed.
     * @type {?}
     */
    NxCheckboxCircleComponent.prototype.checkedChange;
    /**
     * Whether the checkbox circle is checked.
     * @type {?}
     */
    NxCheckboxCircleComponent.prototype.checked;
    /**
     * The name of the icon that is displayed. If not added, no icon will be displayed.
     * @type {?}
     */
    NxCheckboxCircleComponent.prototype.iconName;
    /**
     * If set, the passed SVG will be displayed in the not-checked state.
     * @type {?}
     */
    NxCheckboxCircleComponent.prototype.svg;
    /**
     * If set, the passed SVG will be displayed in the checked state.
     * @type {?}
     */
    NxCheckboxCircleComponent.prototype.svgChecked;
    /**
     * Label to be displayed under circle.
     * @type {?}
     */
    NxCheckboxCircleComponent.prototype.label;
    /**
     * @type {?}
     * @private
     */
    NxCheckboxCircleComponent.prototype.onChangeCallback;
    /**
     * @type {?}
     * @private
     */
    NxCheckboxCircleComponent.prototype.onTouchedCallback;
}
