// @allianz/ngx-ndbx/esm5/pagination/pagination-texts.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/**
 * Interface for providing custom labels in a simple pagination.
 * With an implementation you can localize your pagination with the NX_PAGINATION_TEXTS injection token.
 * @record
 */
export function IPaginationTexts() { }
if (false) {
    /**
     * Label that should replace 'previous'.
     * @type {?}
     */
    IPaginationTexts.prototype.previous;
    /**
     * Label that should replace 'next'.
     * @type {?}
     */
    IPaginationTexts.prototype.next;
    /**
     * Label that should replace 'first'.
     *
     * Optional attribute needed only for the advanced pagination.
     * @type {?|undefined}
     */
    IPaginationTexts.prototype.first;
    /**
     * Label that should replace 'last'.
     *
     * Optional attribute needed only for the advanced pagination.
     * @type {?|undefined}
     */
    IPaginationTexts.prototype.last;
    /**
     * Label that should replace 'of'.
     * @type {?}
     */
    IPaginationTexts.prototype.ofLabel;
    /**
     * Label that should replace the aria label.
     * @type {?}
     */
    IPaginationTexts.prototype.ariaLabel;
}
/** @type {?} */
export var DefaultPaginationTexts = {
    previous: 'Previous',
    next: 'Next',
    first: 'First',
    last: 'Last',
    ofLabel: 'of',
    ariaLabel: 'Please select your page'
};
/**
 * InjectionToken for pagination that can be used to override default locale code.
 * @type {?}
 */
export var NX_PAGINATION_TEXTS = new InjectionToken('nx-pagination-texts');
